---
title: 'Mendelian Randomization: <trait> and Alzheimers disease'
author: "Shea Andrews"
date: "`r Sys.Date()`"
output:
  html_document:
    df_print: paged
    toc: true
    number_sections: true
    toc_depth: 4
    toc_float:
      collapsed: false
      smooth_scroll: false
editor_options:
  chunk_output_type: console
params:
  load.summary:
  aaos.summary:
  ab42.summary:
  ptau.summary:
  tau.summary:
  hipv.summary:
  trait.summary:
  load.clump:
  aaos.clump:
  ab42.clump:
  ptau.clump:
  tau.clump:
  hipv.clump:
  trait.name:
  trait.blurb:
  trait.code:
  out.path:
  p.threshold: 
---

## Data sources
`r params$trait.blurb`

***Late Onset Alzheimer's disease*** (LOAD, Lambert et al 2013): The International Genomics of Alzheimer’s Project (IGAP) is a meta-analysis of 4 previously published GWAS datasets: the European Alzheimer’s Disease Imitative (EADI), the Alzheimer Disease Genetics Consortium (ADGC), Cohorts for Heart and Aging Research in Genomic Epidemiology (CHARGE), and Genetic and Environmental Risk in AD (GERAD) and includes a sample of 17,008 LOAD cases and 37,154 cognitively normal elder controls. Participants in IGAP were of European ancestry, the average age was 71 and 58.4% of participants were women.


***Alzheimer's Age of Onset Surivial*** (AAOS, Huang et al 2017): A GWAS of age of onset in LOAD was conducted in 14,406 AD case samples and 25,849 control samples from the IGAP using Cox proportional hazard regressions. Participants were of European ancestry, in cases the the average AAO was 74.8 and 61.7% were women, in controls the average AAE was 79.0 and 59.6% were women.

***Hippocampal Volume*** [Hibar et al 2017](https://doi.org/10.1038/ncomms13624): A GWAS of hippocampal volume perfomed in 26,814 (ENIGMA and CHARGE consortiums) individules of European Ancestry discovered 9 independent loci.

`r if(params$p.threshold != '5e-6') {"\\begin{comment}"}`
***CSF Ab42, tau & ptau*** (AB42, ptau, tau, Deming et al 2017): A GWAS of CSF AB42, ptau and tau levels (pg/mL) was conducted in 3,146 participants. Participants were of Eurpean ancestry.
`r if(params$p.threshold != '5e-6') {"\\end{comment}"}`

```{r setup, include=FALSE, echo=FALSE}
knitr::opts_chunk$set(echo = TRUE)

### ===== Load packages ===== ###

suppressMessages(library(tidyverse))   ## For data wrangling
suppressMessages(library(Hmisc))       ## Contains miscillaneous funtions
suppressMessages(library(TwoSampleMR)) ## For conducting MR https://mrcieu.github.io/TwoSampleMR/
suppressMessages(library(haploR))


##  ---- Find Proxy SNPs - uses haploR ---- ##
FindProxys <-function(snplist, dat){
  ProxySnps <- queryHaploreg(query = snplist, ldThresh = 0.8,
              ldPop = "EUR", timeout = 100, encoding = "UTF-8", verbose = TRUE) %>% 
  select(query_snp_rsid, rsID, chr, pos_hg38, r2, `D'`, is_query_snp, ref, alt) %>% 
  rename(ld.r2 = r2, Dprime = `D'`)

query_snps <- ProxySnps %>% filter(is_query_snp == 1) %>% select(query_snp_rsid, ref, alt)

proxy.snps <- left_join(ProxySnps, dat, by = c('rsID' = 'SNP')) %>% 
  filter(!is.na(Effect_allele)) %>%                       ## remove snps with missing information
  filter(is_query_snp == 0) %>%                           ## remove query snps
  filter(nchar(ref) == 1) %>% filter(nchar(alt) == 1) %>% ## remove indels
  group_by(query_snp_rsid) %>%                            ## by query snp
  arrange(-ld.r2) %>%                                     ## arrange by ld
  slice(1) %>%                                            ## select top proxy snp
  ungroup() %>% 
  rename(Effect_allele.proxy = Effect_allele, Non_Effect_allele.proxy = Non_Effect_allele) %>% ## rename alleles
  select(-chr, -pos_hg38, -is_query_snp) %>%              ## remove uneeded columns
  ## Join proxy snps to query snps
  left_join(query_snps, by = 'query_snp_rsid', suffix = c('.proxy', "")) %>% 
  ## for each proxy snp, coded correlated allels
  mutate(Effect_allele = ifelse(Effect_allele.proxy == ref.proxy, ref, alt)) %>%
  mutate(Non_Effect_allele = ifelse(Non_Effect_allele.proxy == ref.proxy, ref, alt)) %>% 
  rename(SNP = query_snp_rsid) 

proxy.snps
}

```

```{r datain_igap, include=FALSE, echo=FALSE, cache=T}
### ===== Load Summary Statisitics ===== ###
## Assumes summayr statsticsts have the following column names:
#   1. SNP
#   2. CHR
#   3. POS
#   4. Effect_allele
#   5. Non_Effect_allele
#   6. EAF
#   7. Beta
#   8. SE
#   9. P
#   10. r2
#   11. N
## ---- Trait ---- ##
trait.dat <- read_tsv(params$trait.summary)

## ---- Alzheimer's disease ---- ##
#  LOAD: Lambert et al 2013
load.dat <- read_tsv(params$load.summary) %>%
  filter(P < as.numeric(params$p.threshold))
#  AAOS: Huang et al 2017
aaos.dat <- read_tsv(params$aaos.summary) %>%
  filter(P < as.numeric(params$p.threshold))

#  Hippocampul Volume
hipv.dat <- read_tsv(params$hipv.summary) %>%
  filter(P < as.numeric(params$p.threshold))

## ---- Clumped SNPs ---- ##
load_clump <- read_table2(params$load.clump) %>%
  filter(!is.na(CHR)) %>%
  select(CHR, F, SNP, BP, P, TOTAL, NSIG)%>%
  filter(P < as.numeric(params$p.threshold))

aaos_clump <- read_table2(params$aaos.clump) %>%
  filter(!is.na(CHR)) %>%
  select(CHR, F, SNP, BP, P, TOTAL, NSIG) %>%
  filter(P < as.numeric(params$p.threshold))

hipv_clump <- read_table2(params$hipv.clump) %>%
  filter(!is.na(CHR)) %>%
  select(CHR, F, SNP, BP, P, TOTAL, NSIG) %>%
  filter(P < as.numeric(params$p.threshold))

```

`r if(params$p.threshold != '5e-6') {"\\begin{comment}"}`
```{r datain_csf, include=FALSE, echo=FALSE, cache=T, eval=params$p.threshold == '5e-6'}
## ---- CSF ---- ##
#  csf: Demming et al 2017

ab42.dat <- read_tsv(params$ab42.summary) %>%
    filter(P < as.numeric(params$p.threshold))
ptau.dat <- read_tsv(params$ptau.summary) %>%
    filter(P < as.numeric(params$p.threshold))
tau.dat <- read_tsv(params$tau.summary) %>%
    filter(P < as.numeric(params$p.threshold))

ab42_clump <- read_table2(params$ab42.clump) %>%
    filter(!is.na(CHR)) %>%
    select(CHR, F, SNP, BP, P, TOTAL, NSIG) %>%
    filter(P < as.numeric(params$p.threshold))

ptau_clump <- read_table2(params$ptau.clump) %>%
    filter(!is.na(CHR)) %>%
    select(CHR, F, SNP, BP, P, TOTAL, NSIG) %>%
    filter(P < as.numeric(params$p.threshold))

tau_clump <- read_table2(params$tau.clump) %>%
    filter(!is.na(CHR)) %>%
    select(CHR, F, SNP, BP, P, TOTAL, NSIG) %>%
    filter(P < as.numeric(params$p.threshold))
```
`r if(params$p.threshold != '5e-6') {"\\end{comment}"}`

## Instrumental Vaibles
**LD Clumping:** For standard two sample MR it is important to ensure that the instruments for the exposure are independent. LD clumping can be performed using the data_clump function from the TwoSampleMR package, which uses EUR samples from the 1000 genomes project to estimate LD between SNPs and amonst SNPs that have and LD above a given threshold, only the SNP with the lowest p-value will be retained.
<br>

**Proxy SNPs:** SNPs associated with LOAD, AAOS, AB42, ptau and tau were extracted from the GWAS of `r params$trait.name`. Where SNPs were not available in the outcome GWAS, the EUR thousand genomes was queried to identified potential proxy SNPs that are in linkage disequilibrium (r2 > 0.8) of the missing SNP.


### LOAD
```{r LOAD_clump, echo=F, message=F, warning=F, cache=F}
load <- semi_join(load.dat, load_clump, by = 'SNP')
```

LOAD: `r nrow(load.dat)` SNPs were associated with LOAD at p < `r params$p.threshold`, with `r nrow(load_clump)` SNPs remaining after LD clumping (Table 1).
<br>

**Table 1:** SNPS associated with LOAD
```{r LOAD_table, echo=FALSE}
as.data.frame(load)
```
<br>

```{r data_joining trait_load, warning=FALSE, echo=FALSE, cache=F}
# LOAD
dat.load.trait <- trait.dat %>%
  right_join(select(load, SNP), by = 'SNP')

miss.load.trait <- filter(dat.load.trait, is.na(CHR))

# If snps are missing, find proxy snp
if(nrow(miss.load.trait) >= 1){
  # Search for Proxy SNPs
  proxy.load.trait <- FindProxys(miss.load.trait$SNP, trait.dat)

  # Combine proxy snps with trait data
  dat.load.trait <- dat.load.trait %>%
    filter(SNP %nin% miss.load.trait$SNP) %>%
    bind_rows(select(proxy.load.trait, SNP, CHR, POS, Effect_allele, Non_Effect_allele, EAF, Beta, SE, P, r2)) %>%
    arrange(CHR, POS)
}

```

Of the the `r nrow(load)` SNPs associated with LOAD, `r nrow(load) - nrow(miss.load.trait)` were available in the `r params$trait.name` GWAS (Table 2).

**Table 2:** SNPS associated with `r params$trait.name` avalible in LOAD GWAS
```{r trait_LOAD_table, echo=FALSE}
as.data.frame(dat.load.trait)
```
<br>

### AAOS

```{r aaos_clump, echo=F, message=F, warning=F, cache=F}
aaos <- semi_join(aaos.dat, aaos_clump, by = 'SNP')
```

AAOS: `r nrow(aaos.dat)` SNPs were associated with AAOS at p < `r params$p.threshold`, with `r nrow(aaos_clump)` SNPs remaining after LD clumping (Table 3).
<br>

**Table 3:** SNPS associated with AAOS
```{r AAOS_table, echo=FALSE}
as.data.frame(aaos)
```
<br>

```{r data_joining trait_aaos, warning=FALSE, echo=FALSE, cache=F}
# AAOS
dat.aaos.trait <- trait.dat %>%
  right_join(select(aaos, SNP), by = 'SNP')

miss.aaos.trait <- filter(dat.aaos.trait, is.na(CHR))

# If snps are missing, find proxy snp
if(nrow(miss.aaos.trait) >= 1){
  # Search for Proxy SNPs
  proxy.aaos.trait <- FindProxys(miss.aaos.trait$SNP, trait.dat)

  # Combine proxy snps with trait data
  dat.aaos.trait <- dat.aaos.trait %>%
    filter(SNP %nin% miss.aaos.trait$SNP) %>%
    bind_rows(select(proxy.aaos.trait, SNP, CHR, POS, Effect_allele, Non_Effect_allele, EAF, Beta, SE, P, r2)) %>%
    arrange(CHR, POS)
}

```

Of the the `r nrow(aaos)` SNPs associated with AAOS, `r nrow(aaos) - nrow(miss.aaos.trait)` were available in the `r params$trait.name` GWAS (Table 4).

**Table 4:** SNPS associated with `r params$trait.name` avalible in AAOS GWAS
```{r trait_AAOS_table, echo=FALSE}
as.data.frame(dat.aaos.trait)
```
<br>

### Hippocampul volume
```{r hipv_clump, echo=F, message=F, warning=F, cache=F}
hipv <- semi_join(hipv.dat, hipv_clump, by = 'SNP')
```

hipv: `r nrow(hipv.dat)` SNPs were associated with hipv at p < `r params$p.threshold`, with `r nrow(hipv_clump)` SNPs remaining after LD clumping (Table 1).
<br>
  
  **Table 1:** SNPS associated with hipv
```{r hipv_table, echo=FALSE}
as.data.frame(hipv)
```
<br>
  
  ```{r data_joining trait_hipv, warning=FALSE, echo=FALSE, cache=F}
# hipv
dat.hipv.trait <- trait.dat %>%
  right_join(select(hipv, SNP), by = 'SNP')

miss.hipv.trait <- filter(dat.hipv.trait, is.na(CHR))

# If snps are missing, find proxy snp
if(nrow(miss.hipv.trait) >= 1){
  # Search for Proxy SNPs
  proxy.hipv.trait <- FindProxys(miss.hipv.trait$SNP, trait.dat)
  
  # Combine proxy snps with trait data
  dat.hipv.trait <- dat.hipv.trait %>%
    filter(SNP %nin% miss.hipv.trait$SNP) %>%
    bind_rows(select(proxy.hipv.trait, SNP, CHR, POS, Effect_allele, Non_Effect_allele, EAF, Beta, SE, P, r2)) %>%
    arrange(CHR, POS)
}

```

Of the the `r nrow(hipv)` SNPs associated with hipv, `r nrow(hipv) - nrow(miss.hipv.trait)` were available in the `r params$trait.name` GWAS (Table 2).

**Table 2:** SNPS associated with `r params$trait.name` avalible in hipv GWAS
```{r trait_hipv_table, echo=FALSE}
as.data.frame(dat.hipv.trait)
```
<br>

`r if(params$p.threshold != '5e-6') {"\\begin{comment}"}`
### AB42

```{r ab42_clump, echo=F, message=F, warning=F, cache=F, eval=params$p.threshold == '5e-6'}
ab42 <- semi_join(ab42.dat, ab42_clump, by = 'SNP')
```

```{r, eval=params$p.threshold == '5e-6', results='asis', echo=F}
paste('AB42:', nrow(ab42.dat), 'SNPs were associated with AB42 at p <', params$p.threshold, 'with', nrow(ab42_clump), 'SNPs remaining after LD clumping (Table 5).')
```
<br>

**Table 5:** SNPS associated with AB42
```{r AB42_table, echo=FALSE, eval=params$p.threshold == '5e-6'}
as.data.frame(ab42)
```
<br>

```{r data_joining trait_ab42, warning=FALSE, echo=FALSE, cache=F, eval=params$p.threshold == '5e-6'}
# AB42
dat.ab42.trait <- trait.dat %>%
  right_join(select(ab42, SNP), by = 'SNP')

miss.ab42.trait <- filter(dat.ab42.trait, is.na(CHR))

# If snps are missing, find proxy snp
if(nrow(miss.ab42.trait) >= 1){
  # Search for Proxy SNPs
  proxy.ab42.trait <- FindProxys(miss.ab42.trait$SNP, trait.dat)

  # Combine proxy snps with trait data
  dat.ab42.trait <- dat.ab42.trait %>%
    filter(SNP %nin% miss.ab42.trait$SNP) %>%
    bind_rows(select(proxy.ab42.trait, SNP, CHR, POS, Effect_allele, Non_Effect_allele, EAF, Beta, SE, P, r2)) %>%
    arrange(CHR, POS)
}

```

```{r, eval=params$p.threshold == '5e-6', results='asis', echo=F}
paste('Of the the', nrow(ab42), 'SNPs associated with AB42,', (nrow(ab42) - nrow(miss.ab42.trait)), 'were available in the', params$trait.name, 'GWAS (Table 6).')
```


**Table 6:** SNPS associated with `r params$trait.name` avalible in AB42 GWAS
```{r trait_AB42_table, echo=FALSE, eval=params$p.threshold == '5e-6'}
as.data.frame(dat.ab42.trait)
```
<br>

### Ptau

```{r ptau_clump, echo=F, message=F, warning=F, cache=F, eval=params$p.threshold == '5e-6'}
ptau <- semi_join(ptau.dat, ptau_clump, by = 'SNP')
```

```{r, eval=params$p.threshold == '5e-6', results='asis', echo=F}
paste('Ptau:', nrow(ptau.dat), 'SNPs were associated with Ptau at p <', params$p.threshold, 'with', nrow(ptau_clump), 'SNPs remaining after LD clumping (Table 7).')
```
<br>

**Table 7:** SNPS associated with Ptau
```{r Ptau_table, echo=FALSE, eval=params$p.threshold == '5e-6'}
as.data.frame(ptau)
```
<br>

```{r data_joining trait_ptau, warning=FALSE, echo=FALSE, cache=F, eval=params$p.threshold == '5e-6'}
# Ptau
dat.ptau.trait <- trait.dat %>%
  right_join(select(ptau, SNP), by = 'SNP')

miss.ptau.trait <- filter(dat.ptau.trait, is.na(CHR))

# If snps are missing, find proxy snp
if(nrow(miss.ptau.trait) >= 1){
  # Search for Proxy SNPs
  proxy.ptau.trait <- FindProxys(miss.ptau.trait$SNP, trait.dat)

  # Combine proxy snps with trait data
  dat.ptau.trait <- dat.ptau.trait %>%
    filter(SNP %nin% miss.ptau.trait$SNP) %>%
    bind_rows(select(proxy.ptau.trait, SNP, CHR, POS, Effect_allele, Non_Effect_allele, EAF, Beta, SE, P, r2)) %>%
    arrange(CHR, POS)
}

```

```{r, eval=params$p.threshold == '5e-6', results='asis', echo=F}
paste('Of the the', nrow(ptau), 'SNPs associated with Ptau,', (nrow(ptau) - nrow(miss.ptau.trait)), 'were available in the', params$trait.name, 'GWAS (Table 6).')
```


**Table 8:** SNPS associated with `r params$trait.name` avalible in Ptau GWAS
```{r trait_Ptau_table, echo=FALSE, eval=params$p.threshold == '5e-6'}
as.data.frame(dat.ptau.trait)
```
<br>

### Tau

```{r tau_clump, echo=F, message=F, warning=F, cache=F, eval=params$p.threshold == '5e-6'}
tau <- semi_join(tau.dat, tau_clump, by = 'SNP')
```

```{r, eval=params$p.threshold == '5e-6', results='asis', echo=F}
paste('Tau:', nrow(tau.dat), 'SNPs were associated with Tau at p <', params$p.threshold, 'with', nrow(tau_clump), 'SNPs remaining after LD clumping (Table 9).')
```
<br>

**Table 9:** SNPS associated with Tau
```{r Tau_table, echo=FALSE, eval=params$p.threshold == '5e-6'}
as.data.frame(tau)
```
<br>

```{r data_joining trait_tau, warning=FALSE, echo=FALSE, cache=F, eval=params$p.threshold == '5e-6'}
# Tau
dat.tau.trait <- trait.dat %>%
  right_join(select(tau, SNP), by = 'SNP')

miss.tau.trait <- filter(dat.tau.trait, is.na(CHR))

# If snps are missing, find proxy snp
if(nrow(miss.tau.trait) >= 1){
  # Search for Proxy SNPs
  proxy.tau.trait <- FindProxys(miss.tau.trait$SNP, trait.dat)

  # Combine proxy snps with trait data
  dat.tau.trait <- dat.tau.trait %>%
    filter(SNP %nin% miss.tau.trait$SNP) %>%
    bind_rows(select(proxy.tau.trait, SNP, CHR, POS, Effect_allele, Non_Effect_allele, EAF, Beta, SE, P, r2)) %>%
    arrange(CHR, POS)
}

```

```{r, eval=params$p.threshold == '5e-6', results='asis', echo=F}
paste('Of the the', nrow(tau), 'SNPs associated with Tau,', (nrow(tau) - nrow(miss.tau.trait)), 'were available in the', params$trait.name, 'GWAS (Table 6).')
```


**Table 10:** SNPS associated with `r params$trait.name` avalible in Tau GWAS
```{r trait_Tau_table, echo=FALSE, eval=params$p.threshold == '5e-6'}
as.data.frame(dat.tau.trait)
```
<br>

`r if(params$p.threshold != '5e-6') {"\\end{comment}"}`

## Data harmonization
Harmonize the exposure and outcome datasets so that the effect of a SNP on the exposure and the effect of that SNP on the outcome correspond to the same allele. The harmonise_data function from the TwoSampleMR package can be used to perform the harmonization step, by default it try's to infer the forward strand alleles using allele frequency information. EAF were not availbe in the IGAP summary statisitics, as such the allele frequencies reported in the AAOS anaylsis were used.
<br>

### LOAD ~ `r params$trait.name`
```{r harmonize_trait_load, warning=FALSE, echo=FALSE, include=FALSE, cache=F}
# Format LOAD
## ---- Exposure ---- ##
mr_load <- format_data(load, type = 'exposure',
                            phenotype_col = 'LOAD',
                            snp_col = 'SNP',
                            beta_col = "Beta",
                            se_col = "SE",
                            eaf_col = "EAF",
                            effect_allele_col = "Effect_allele",
                            other_allele_col = "Non_Effect_allele",
                            pval_col = "P")
mr_load <- mutate(mr_load, exposure = 'LOAD')

## Format Trait
mr_load.trait.dat <- format_data(dat.load.trait, type = 'outcome',
                          phenotype_col = 'AD',
                          snp_col = 'SNP',
                          beta_col = "Beta",
                          se_col = "SE",
                          eaf_col = "EAF",
                          effect_allele_col = "Effect_allele",
                          other_allele_col = "Non_Effect_allele",
                          pval_col = "P")
mr_load.trait.dat <- mutate(mr_load.trait.dat, outcome = params$trait.code)

# harmonize LOAD
load_trait.MRdat <- harmonise_data(mr_load, mr_load.trait.dat)
```

**Table 11:** Harmonized `r params$trait.name` and LOAD datasets
```{r trait_load, echo=FALSE, cache=F}
as.data.frame(load_trait.MRdat %>%
  select(SNP, effect_allele.exposure, other_allele.exposure, effect_allele.outcome, other_allele.outcome, beta.exposure, beta.outcome, eaf.exposure, eaf.outcome, se.outcome, pval.outcome, se.exposure, pval.exposure))
```
<br>

### AAOS ~ `r params$trait.name`
```{r harmonize_trait_aaos, warning=FALSE, echo=FALSE, include=FALSE, cache=F}
# Format AAOS
## ---- Exposure ---- ##
mr_aaos <- format_data(aaos, type = 'exposure',
                       phenotype_col = 'AAOS',
                       snp_col = 'SNP',
                       beta_col = "Beta",
                       se_col = "SE",
                       eaf_col = "EAF",
                       effect_allele_col = "Effect_allele",
                       other_allele_col = "Non_Effect_allele",
                       pval_col = "P")
mr_aaos <- mutate(mr_aaos, exposure = 'AAOS')

## Format Trait
mr_aaos.trait.dat <- format_data(dat.aaos.trait, type = 'outcome',
                                 phenotype_col = 'AD',
                                 snp_col = 'SNP',
                                 beta_col = "Beta",
                                 se_col = "SE",
                                 eaf_col = "EAF",
                                 effect_allele_col = "Effect_allele",
                                 other_allele_col = "Non_Effect_allele",
                                 pval_col = "P")
mr_aaos.trait.dat <- mutate(mr_aaos.trait.dat, outcome = params$trait.code)

# harmonize
aaos_trait.MRdat <- harmonise_data(mr_aaos, mr_aaos.trait.dat)
```

**Table 12:** Harmonized `r params$trait.name` and AAOS datasets
```{r trait_aaos, echo=FALSE, cache=F}
as.data.frame(aaos_trait.MRdat %>%
                select(SNP, effect_allele.exposure, other_allele.exposure, effect_allele.outcome, other_allele.outcome, beta.exposure, beta.outcome, eaf.exposure, eaf.outcome, se.outcome, pval.outcome, se.exposure, pval.exposure))
```
<br>

### Hippcampul Volume ~ `r params$trait.name`
```{r harmonize_trait_hipv, warning=FALSE, echo=FALSE, include=FALSE, cache=F}
## ---- Exposure ---- ##
mr_hipv <- format_data(hipv, type = 'exposure',
                       phenotype_col = 'HV',
                       snp_col = 'SNP',
                       beta_col = "Beta",
                       se_col = "SE",
                       eaf_col = "EAF",
                       effect_allele_col = "Effect_allele",
                       other_allele_col = "Non_Effect_allele",
                       pval_col = "P")
mr_hipv <- mutate(mr_hipv, exposure = 'HV')

## Format Trait
mr_hipv.trait.dat <- format_data(dat.hipv.trait, type = 'outcome',
                                 phenotype_col = 'AD',
                                 snp_col = 'SNP',
                                 beta_col = "Beta",
                                 se_col = "SE",
                                 eaf_col = "EAF",
                                 effect_allele_col = "Effect_allele",
                                 other_allele_col = "Non_Effect_allele",
                                 pval_col = "P")
mr_hipv.trait.dat <- mutate(mr_hipv.trait.dat, outcome = params$trait.code)

# harmonize hipv
hipv_trait.MRdat <- harmonise_data(mr_hipv, mr_hipv.trait.dat)
```

**Table 11:** Harmonized `r params$trait.name` and HV datasets
```{r trait_hipv, echo=FALSE, cache=F}
as.data.frame(hipv_trait.MRdat %>%
                select(SNP, effect_allele.exposure, other_allele.exposure, effect_allele.outcome, other_allele.outcome, beta.exposure, beta.outcome, eaf.exposure, eaf.outcome, se.outcome, pval.outcome, se.exposure, pval.exposure))
```
<br>
  
`r if(params$p.threshold != '5e-6') {"\\begin{comment}"}`
### AB42 ~ `r params$trait.name`
```{r harmonize_trait_ab42, warning=FALSE, echo=FALSE, include=FALSE, cache=F, eval=params$p.threshold == '5e-6'}
# Format AB42
## ---- Exposure ---- ##
mr_ab42 <- format_data(ab42, type = 'exposure',
                       phenotype_col = 'AB42',
                       snp_col = 'SNP',
                       beta_col = "Beta",
                       se_col = "SE",
                       eaf_col = "EAF",
                       effect_allele_col = "Effect_allele",
                       other_allele_col = "Non_Effect_allele",
                       pval_col = "P")
mr_ab42 <- mutate(mr_ab42, exposure = 'AB42')

## Format Trait
mr_ab42.trait.dat <- format_data(dat.ab42.trait, type = 'outcome',
                                 phenotype_col = 'AD',
                                 snp_col = 'SNP',
                                 beta_col = "Beta",
                                 se_col = "SE",
                                 eaf_col = "EAF",
                                 effect_allele_col = "Effect_allele",
                                 other_allele_col = "Non_Effect_allele",
                                 pval_col = "P")
mr_ab42.trait.dat <- mutate(mr_ab42.trait.dat, outcome = params$trait.code)

# harmonize
ab42_trait.MRdat <- harmonise_data(mr_ab42, mr_ab42.trait.dat)
```

**Table 13:** Harmonized `r params$trait.name` and AB42 datasets
```{r trait_ab42, echo=FALSE, cache=F, eval=params$p.threshold == '5e-6'}
as.data.frame(ab42_trait.MRdat %>%
                select(SNP, effect_allele.exposure, other_allele.exposure, effect_allele.outcome, other_allele.outcome, beta.exposure, beta.outcome, eaf.exposure, eaf.outcome, se.outcome, pval.outcome, se.exposure, pval.exposure))
```
<br>

### Ptau ~ `r params$trait.name`
```{r harmonize_trait_ptau, warning=FALSE, echo=FALSE, include=FALSE, cache=F, eval=params$p.threshold == '5e-6'}
# Format Ptau
## ---- Exposure ---- ##
mr_ptau <- format_data(ptau, type = 'exposure',
                       phenotype_col = 'Ptau',
                       snp_col = 'SNP',
                       beta_col = "Beta",
                       se_col = "SE",
                       eaf_col = "EAF",
                       effect_allele_col = "Effect_allele",
                       other_allele_col = "Non_Effect_allele",
                       pval_col = "P")
mr_ptau <- mutate(mr_ptau, exposure = 'Ptau')

## Format Trait
mr_ptau.trait.dat <- format_data(dat.ptau.trait, type = 'outcome',
                                 phenotype_col = 'AD',
                                 snp_col = 'SNP',
                                 beta_col = "Beta",
                                 se_col = "SE",
                                 eaf_col = "EAF",
                                 effect_allele_col = "Effect_allele",
                                 other_allele_col = "Non_Effect_allele",
                                 pval_col = "P")
mr_ptau.trait.dat <- mutate(mr_ptau.trait.dat, outcome = params$trait.code)

# harmonize
ptau_trait.MRdat <- harmonise_data(mr_ptau, mr_ptau.trait.dat)
```

**Table 14:** Harmonized `r params$trait.name` and Ptau datasets
```{r trait_ptau, echo=FALSE, cache=F, eval=params$p.threshold == '5e-6'}
as.data.frame(ptau_trait.MRdat %>%
                select(SNP, effect_allele.exposure, other_allele.exposure, effect_allele.outcome, other_allele.outcome, beta.exposure, beta.outcome, eaf.exposure, eaf.outcome, se.outcome, pval.outcome, se.exposure, pval.exposure))
```
<br>

### Tau ~ `r params$trait.name`
```{r harmonize_trait_tau, warning=FALSE, echo=FALSE, include=FALSE, cache=F, eval=params$p.threshold == '5e-6'}
# Format Tau
## ---- Exposure ---- ##
mr_tau <- format_data(tau, type = 'exposure',
                       phenotype_col = 'Tau',
                       snp_col = 'SNP',
                       beta_col = "Beta",
                       se_col = "SE",
                       eaf_col = "EAF",
                       effect_allele_col = "Effect_allele",
                       other_allele_col = "Non_Effect_allele",
                       pval_col = "P")
mr_tau <- mutate(mr_tau, exposure = 'Tau')

## Format Trait
mr_tau.trait.dat <- format_data(dat.tau.trait, type = 'outcome',
                                 phenotype_col = 'AD',
                                 snp_col = 'SNP',
                                 beta_col = "Beta",
                                 se_col = "SE",
                                 eaf_col = "EAF",
                                 effect_allele_col = "Effect_allele",
                                 other_allele_col = "Non_Effect_allele",
                                 pval_col = "P")
mr_tau.trait.dat <- mutate(mr_tau.trait.dat, outcome = params$trait.code)

# harmonize
tau_trait.MRdat <- harmonise_data(mr_tau, mr_tau.trait.dat)
```

**Table 15:** Harmonized `r params$trait.name` and Tau datasets
```{r trait_tau, echo=FALSE, cache=F, eval=params$p.threshold == '5e-6'}
as.data.frame(tau_trait.MRdat %>%
                select(SNP, effect_allele.exposure, other_allele.exposure, effect_allele.outcome, other_allele.outcome, beta.exposure, beta.outcome, eaf.exposure, eaf.outcome, se.outcome, pval.outcome, se.exposure, pval.exposure))
```
<br>
`r if(params$p.threshold != '5e-6') {"\\end{comment}"}`

```{r write, echo=FALSE, message=F, warning=F, include=F}
write_csv(load_trait.MRdat, paste0(params$out.path, params$trait.code, '/', 'load_', params$trait.code, '_', params$p.threshold, '_BMRdat.csv'))
write_csv(aaos_trait.MRdat, paste0(params$out.path, params$trait.code, '/', 'aaos_', params$trait.code, '_', params$p.threshold, '_BMRdat.csv'))
write_csv(hipv_trait.MRdat, paste0(params$out.path, params$trait.code, '/', 'hipv_', params$trait.code, '_', params$p.threshold, '_BMRdat.csv'))

if(params$p.threshold == '5e-6'){
  write_csv(ab42_trait.MRdat, paste0(params$out.path, params$trait.code, '/', 'ab42_', params$trait.code, '_', params$p.threshold, '_BMRdat.csv'))
  write_csv(ptau_trait.MRdat, paste0(params$out.path, params$trait.code, '/', 'ptau_', params$trait.code, '_', params$p.threshold, '_BMRdat.csv'))
  write_csv(tau_trait.MRdat, paste0(params$out.path, params$trait.code, '/', 'tau_', params$trait.code, '_', params$p.threshold, '_BMRdat.csv'))
}
```
\
