---
title: "Mendelian Randomization Analysis"
author: "Dr. Shea Andrews"
date: "`r Sys.Date()`"
output:
  html_document:
    df_print: paged
    keep_md: false
    toc: true
    number_sections: false
    toc_depth: 4
    toc_float:
      collapsed: false
      smooth_scroll: false
params:
  rwd: rwd
  traits: traits
  exposure.code: exposure.code
  exposure.snps: exposure.snps
  outcome.code: outcome.code
  outcome.snps: outcome.snps
  proxy.snps: proxy.snps
  harmonized.dat: harmonized.dat
  p.threshold: p.threshold
  mrpresso_global: mrpresso_global
  out: out
editor_options: 
  chunk_output_type: console
---
```{r test, include=FALSE, eval= FALSE}
## For testing 
rwd <- 'out'
out <- 'out'
exposure.code <- 'mvpa'
outcome.code <- 'load'
p.threshold <- '5e-8'

traits <- '~/Dropbox/Research/PostDoc-MSSM/2_MR/1_RawData/MRTraits.csv'

exposure.snps <- paste0('~/Dropbox/Research/PostDoc-MSSM/2_MR/2_DerivedData/', exposure.code, '/', exposure.code, '_', p.threshold, '_SNPs.txt')
outcome.snps <- paste0('~/Dropbox/Research/PostDoc-MSSM/2_MR/2_DerivedData/', exposure.code, '/', outcome.code, '/', 
                       exposure.code, '_', p.threshold, '_', outcome.code, '_SNPs.txt')
proxy.snps <- paste0('~/Dropbox/Research/PostDoc-MSSM/2_MR/2_DerivedData/', exposure.code, '/', outcome.code, '/', 
                       exposure.code, '_', p.threshold, '_', outcome.code, '_MatchedProxys.csv')
harmonized.dat <- paste0('~/Dropbox/Research/PostDoc-MSSM/2_MR/2_DerivedData/', exposure.code, '/', outcome.code, '/', 
                       exposure.code, '_', p.threshold, '_', outcome.code, '_mrpresso_MRdat.csv')
mrpresso_global <- paste0('~/Dropbox/Research/PostDoc-MSSM/2_MR/2_DerivedData/', exposure.code, '/', outcome.code, '/', 
                       exposure.code, '_', p.threshold, '_', outcome.code, '_mrpresso_global.txt')
params <- list(rwd, out, exposure.code, outcome.code, p.threshold, traits, exposure.snps, outcome.snps, proxy.snps, harmonized.dat, mrpresso_global)
names(params) <- c('rwd', 'out', 'exposure.code', 'outcome.code', 'p.threshold', 'traits', 'exposure.snps', 'outcome.snps', 'proxy.snps', 'harmonized.dat', 'mrpresso_global')
```

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_knit$set(root.dir = params$rwd)

library(tidyverse)   ## For data wrangling 
library(TwoSampleMR) ## For conducting MR https://mrcieu.github.io/TwoSampleMR/ 
library(RadialMR)    ## For Radial MR plots  
library(ggplot2)     ## For plotting 
```

```{r traits, include=FALSE}
## Read in trait information
traits <- read_csv(params$traits)

exposure.name <- as.character(traits[grep(paste0("\\b", params$exposure.code, "\\b"), traits$code), 'name'])
exposure.blurb <- as.character(traits[grep(paste0("\\b", params$exposure.code, "\\b"), traits$code), 'blurb'])

outcome.name <- as.character(traits[grep(paste0("\\b", params$outcome.code, "\\b"), traits$code), 'name'])
outcome.blurb <- as.character(traits[grep(paste0("\\b", params$outcome.code, "\\b"), traits$code), 'blurb'])
```

## Instrumental Variables
**SNP Inclusion:** SNPS associated with at a p-value threshold of p < `r params$p.threshold` were included in this analysis.
<br>

**LD Clumping:** For standard two sample MR it is important to ensure that the instruments for the exposure are independent. LD clumping was performed in PLINK using the EUR samples from the 1000 Genomes Project to estimate LD between SNPs, and, amongst SNPs that have an LD above a given threshold, retain only the SNP with the lowest p-value. A clumping window of 250kb, r2 of 0.1 and significance level of 1 was used for the index and secondary SNPs.
<br>

**Proxy SNPs:** Where SNPs were not available in the outcome GWAS, the EUR thousand genomes was queried to identify potential proxy SNPs that are in linkage disequilibrium (r2 > 0.8) of the missing SNP.
<br> 

### Exposure: `r exposure.name`
`r exposure.blurb`
<br> 

**Table 1:** Independent SNPs associated with `r exposure.name`
```{r exposure_table, echo=FALSE, message=F}
exposure <- read_tsv(params$exposure.snps)
as.data.frame(exposure)
```
<br>

### Outcome: `r outcome.name`
`r outcome.blurb`
<br>

**Table 2:** SNPs associated with `r exposure.name` avaliable in `r outcome.name`
```{r outcome_table, echo=FALSE, message=F}
outcome <- read_tsv(params$outcome.snps)
as.data.frame(outcome)
```
<br>

**Table 3:** Proxy SNPs for `r outcome.name`
```{r proxy_table, echo=FALSE, message=F}
proxys <- read_csv(params$proxy.snps)
as.data.frame(proxys)
```
<br>

## Data harmonization
Harmonize the exposure and outcome datasets so that the effect of a SNP on the exposure and the effect of that SNP on the outcome correspond to the same allele. The harmonise_data function from the TwoSampleMR package can be used to perform the harmonization step, by default it try's to infer the forward strand alleles using allele frequency information. EAF were not availbe in the IGAP summary statisitics, as such the allele frequencies reported in the AAOS anaylsis were used.
<br>

**Table 4:** Harmonized `r exposure.name` and `r outcome.name` datasets
```{r harmonized_table, echo=FALSE, message=F}
mrdat <- read_csv(params$harmonized.dat)
as.data.frame(mrdat)
```
<br>

## Instrument Strength
```{r f_stat, echo=F}
## From the RadialMR package.
mrdat <- mrdat %>% 
  mutate(F_stat = beta.exposure^2 / se.exposure^2)
mf <- mrdat %>% filter(mr_keep == TRUE) %>% summarise(F = round(mean(F_stat), 2)) %>% as.numeric()
mf.out <- tibble(id.exposure = mrdat %>% slice(1) %>% select(id.exposure) %>% as.character(), 
                 id.outcome = mrdat %>% slice(1) %>% select(id.outcome) %>% as.character(),
                 outcome = mrdat %>% slice(1) %>% select(outcome) %>% as.character(), 
                 exposure = mrdat %>% slice(1) %>% select(exposure) %>% as.character(), 
                 nsnps = mrdat %>% filter(mr_keep == TRUE) %>% nrow(),
                 mean_F = mf)
```
To ensure that the first assumption of MR is not violated (Non-zero effect assumption), the genetic variants selected should be robustly associated with the exposure. Weak instruments, where the variance in the exposure explained by the the instruments is a small portion of the total variance, may result in poor precission and accuracy of the causal effect estiamte. The strength of an instrument can be evaluated using the F statistic, if F is less than 10, this is an indication of weak instrument.


The mean F of the `r exposure.name` SNPs is `r mf`.


##  MR Results 
To obtain an overall estimate of causal effect, the SNP-exposure and SNP-outcome coefficients were combined in 1) a fixed-effects meta-analysis using an inverse-variance weighted approach (IVW); 2) a Weighted Median approach; 3) Weighted Mode approach and 4) Egger Regression. 


[**IVW**](https://doi.org/10.1002/gepi.21758) is equivalent to a weighted regression of the SNP-outcome coefficients on the SNP-exposure coefficients with the intercept constrained to zero. This method assumes that all variants are valid instrumental variables based on the Mendelian randomization assumptions. The causal estimate of the IVW analysis expresses the causal increase in the outcome (or log odds of the outcome for a binary outcome) per unit change in the exposure. [**Weighted median MR**](https://doi.org/10.1002/gepi.21965) allows for 50% of the instrumental variables to be invalid. [**MR-Egger regression**](https://doi.org/10.1093/ije/dyw220) allows all the instrumental variables to be subject to direct effects (i.e. horizontal pleiotropy), with the intercept representing bias in the causal estimate due to pleiotropy and the slope representing the causal estimate. [**Weighted Mode**](https://doi.org/10.1093/ije/dyx102) gives consistent estimates as the sample size increases if a plurality (or weighted plurality) of the genetic variants are valid instruments.
<br> 

```{r MR_analysis, echo=F, warning=F, message = FALSE}

##  Single snp analysis and meta analysis
res_single <- mr_singlesnp(mrdat,
                           single_method = 'mr_wald_ratio',
                           all_method=c("mr_ivw_fe", "mr_weighted_median", "mr_weighted_mode", "mr_egger_regression"))

res <- mr(mrdat, method_list = c("mr_ivw_fe", "mr_weighted_median", "mr_weighted_mode", "mr_egger_regression"))

```

Table 5 presents the MR causal estimates of genetically predicted `r exposure.name` on `r outcome.name`. 
<br> 

**Table 5** MR causaul estimates for `r exposure.name` on `r outcome.name`
```{r MR_table, echo=FALSE, message=F}
as.data.frame(res)
```
<br> 

Figure 1 illustrates the SNP-specific associations with `r exposure.name` versus the association in `r outcome.name` and the corresponding MR estimates.
<br> 

```{r scatter_plot, echo=F, message = FALSE, warning=F, fig.width=7.5, fig.height=4.7, fig.align='center', fig.cap= 'Fig. 1: Scatterplot of SNP effects for the association of the Exposure on the Outcome'}

##  Scatter Plot
scatter_plot <- mr_scatter_plot(res, mrdat)
scatter_plot[[1]] + theme_bw() + theme(legend.position = "bottom", text = element_text(family="Times", size=12)) + 
  guides(col = guide_legend(nrow = 1)) + scale_colour_discrete() +
  labs(x = paste0('SNP effect on ', exposure.name), 
       y = paste0('SNP effect on ', outcome.name))  

```
<br>


## Pleiotropy 
A Cochrans Q heterogeneity test can be used to formaly assesse for the presence of heterogenity (Table 6), with excessive heterogeneity indicating that there is a meaningful violation of at least one of the MR assumptions.
these assumptions.. 
<br> 

**Table 6:** Heterogenity Tests
```{r heterogeneity, echo=F, warning=F}
heterogenity <- mr_heterogeneity(mrdat, method_list=c("mr_egger_regression", "mr_ivw"))
as.data.frame(heterogenity)
```
<br>

Figure 2 shows a funnel plot displaying the MR estimates of individual genetic variants against their precession. Aysmmetry in the funnel plot may arrise due to some genetic variants having unusally strong effects on the outcome, which is indicative of directional pleiotropy.
<br>

```{r funnel_plot, echo=F, message = FALSE, warning=F, fig.width=7.5, fig.height=4.7, fig.align='center', fig.cap='Fig. 2: Funnel plot of the MR causal estimates against their precession'}
funnel_plot <- mr_funnel_plot(res_single)
funnel_plot[[1]] + theme_bw() + scale_colour_discrete() + 
  theme(legend.position = "bottom", text = element_text(family="Times", size=12))

```
<br>

Figure 3 shows a [Radial Plots](https://github.com/WSpiller/RadialMR) can be used to visualize influential data points with large contributions to Cochran's Q Statistic that may bias causal effect estimates. 

```{r Radial_IVWt, echo = F, cache=F, message=F, warning=F, include=F}
radial.dat <- mrdat %>% filter(mr_keep == TRUE) %>%  
  with(., format_radial(beta.exposure, beta.outcome, se.exposure, se.outcome, SNP))
radial.ivw <- ivw_radial(radial.dat, alpha = 0.05/nrow(radial.dat), weights = 1)
plot_radial(radial.ivw, radial_scale = F)
```

```{r Radial_Plot, echo = F, cache=F, message=F, warning=F, fig.width=7.5, fig.height=4.7, fig.align='center', fig.cap= 'Fig. 4: Radial Plot showing influential data points using Radial IVW'}
plot_radial(radial.ivw, radial_scale = F)
```
<br>

The intercept of the MR-Regression model captures the average pleitropic affect across all genetic variants (Table 7). 
<br> 

**Table 7:** MR Egger test for directional pleitropy
```{r pleitropy, echo=F}
mr_plei <- mr_pleiotropy_test(mrdat)
as.data.frame(mr_plei)
```
<br>

Pleiotropy was also assesed using Mendelian Randomization Pleiotropy RESidual Sum and Outlier [(MR-PRESSO)](https://doi.org/10.1038/s41588-018-0099-7), that allows for the evlation of evaluation of horizontal pleiotropy in a standared MR model (Table 8). MR-PRESSO performs a global test for detection of horizontal pleiotropy and correction of pleiotropy via outlier removal (Table 9). 
<br> 

**Table 8:** MR-PRESSO Global Test for pleitropy
```{r mrpresso, echo = F, cache=F, message=F}
mrpresso_gloabl <- read_tsv(params$mrpresso_global)
as.data.frame(mrpresso_gloabl)
```
<br> 


**Table 9:** MR Estimates after MR-PRESSO outlier removal
```{r mrpresso_res, echo=F, cache=F, message=F}
if(nrow(mrdat) - sum(mrdat$mrpresso_keep, na.rm=TRUE) >= 1){
  plot_outlier = TRUE
  mrdat_mrpresso <- filter(mrdat, mrpresso_keep == T)
  res_mrpresso <- mr(mrdat_mrpresso, method_list = c("mr_ivw_fe", "mr_weighted_median", "mr_weighted_mode", "mr_egger_regression"))
  as.data.frame(res_mrpresso)
}else{
  plot_outlier = FALSE
  res_mrpresso <- data.frame(id.exposure = as.character(mrdat[1,'id.exposure']),
             id.outcome = as.character(mrdat[1,'id.outcome']), 
             outcome = as.character(mrdat[1,'outcome']), 
             exposure = as.character(mrdat[1,'exposure']),
             method = 'mrpresso', 
             nsnp = NA,
             b = NA,
             se = NA,
             pval = NA)
  res_mrpresso
}
```
<br> 

```{r scatter_plot_outlier, eval = plot_outlier, echo=F, message = FALSE, warning=F, fig.width=7.5, fig.height=4.7, fig.align='center', fig.cap= 'Fig. 4: Scatterplot of SNP effects for the association of the Exposure on the Outcome after outlier removal'}

##  Scatter Plot
scatter_plot <- mr_scatter_plot(res_mrpresso, mrdat_mrpresso)
scatter_plot[[1]] + theme_bw() + theme(legend.position = "bottom", text = element_text(family="Times", size=12)) + 
  guides(col = guide_legend(nrow = 1)) + scale_colour_discrete() +
  labs(x = paste0('SNP effect on ', exposure.name), 
       y = paste0('SNP effect on ', outcome.name))  

```

```{r write_out_MR_results, echo=F}
heterogenity %>% mutate(date = Sys.Date()) %>% write_csv(paste0(params$out, '_MR_heterogenity.csv'))
mr_plei %>% mutate(date = Sys.Date()) %>% write_csv(paste0(params$out, '_MR_egger_plei.csv'))
mf.out %>% mutate(date = Sys.Date()) %>% write_csv(paste0(params$out, '_mean_F.csv'))
res %>% mutate(date = Sys.Date()) %>% write_csv(paste0(params$out, '_MR_Results.csv'))
res_mrpresso %>% mutate(date = Sys.Date()) %>% write_csv(paste0(params$out, '_MRPRESSO_Results.csv'))
```
