---
title: "Summary Statistic Cleaning"
author: "Shea Andrews"
date: "7/2/2018"
output: html_document
editor_options: 
  chunk_output_type: console
params: 
  hrc_file: '~/Dropbox/Research/Data/Summary_Statisitics/HRC/HRC_AF.tsv.gz'
  load.summary: '~/Dropbox/Research/Data/Summary_Statisitics/AD/IGAP2013/IGAP_stage_1.txt'
  aaos.summary: '~/Dropbox/Research/Data/Summary_Statisitics/AD/IGAP_AAOS/IGAP_meta1_pos_1filtered.txt'
  ab42.summary: '~/Dropbox/Research/Data/Summary_Statisitics/AD/CSF/Deming_ActaNeuropathol_CSF_AB42_GWAS_summary_statistics.txt.gz'
  ptau.summary: '~/Dropbox/Research/Data/Summary_Statisitics/AD/CSF/Deming_ActaNeuropathol_CSF_ptau_GWAS_summary_statistics.txt.gz'
  tau.summary: '~/Dropbox/Research/Data/Summary_Statisitics/AD/CSF/Deming_ActaNeuropathol_CSF_tau_GWAS_summary_statistics.txt.gz'
  hipv.summary: '~/Dropbox/Research/Data/Summary_Statisitics/Hippocampal_Volume/CHARGE-ENIGMA-HV-METAANALYSIS-201311141.TBL.FINAL.gz'
  alc_cons.summary: '~/Dropbox/Research/Data/Summary_Statisitics/Alcohol/Alc_cons/ALC_CONS_GWAS_FORUKB.txt.gz'
  alc_dep.summary: '~/Dropbox/Research/Data/Summary_Statisitics/Alcohol/Alc_dep/daner_alcdep_eur_jul2017.gz.jul2017.min_n'
  audit.summary: '~/LOAD_minerva/dummy/Data/GwasSummaryStatistics/UKB_AUDIT/AUDIT_UKB_2018_AJP.txt.gz'
  bmi.summary: '~/Dropbox/Research/Data/Summary_Statisitics/BMI/Yengo2018/Meta-analysis_Locke_et_al+UKBiobank_2018_UPDATED.txt.gz'
  smk_ukbb.summary: '~/Dropbox/Research/Data/Summary_Statisitics/Smoking/ukbb/'
  evrsmk.summary: '~/Dropbox/Research/Data/Summary_Statisitics/Smoking/tag/tag.evrsmk.tbl.gz'
  cpd.summary: '~/Dropbox/Research/Data/Summary_Statisitics/Smoking/tag/tag.cpd.tbl.gz'
  mdd.summary: '~/Dropbox/Research/Data/Summary_Statisitics/Depression/Wray2018/PGC_MDD2018_10kSNPs.gz'
  dep.summary: '~/Dropbox/Research/Data/Summary_Statisitics/Depression/Howard2018/UKBiobank_broad_12Jan18.txt.gz'
  diab.summary: '~/Dropbox/Research/Data/Summary_Statisitics/Diabetes/T2D_Meta/Xue_et_al_T2D_META_Nat_Commun_2018.gz'
  mvpa.summary: '~/Dropbox/Research/Data/Summary_Statisitics/Exercise/MVPA_Model1_BOLTLMM_500K.txt.gz'
  acc425.summary: '~/Dropbox/Research/Data/Summary_Statisitics/Exercise/Acc425_Model1_BOLTLMM_500K.txt.gz'
  accave.summary: '~/Dropbox/Research/Data/Summary_Statisitics/Exercise/AccAve_Model1_BOLTLMM_500K.txt.gz'
  ssoe.summary: '~/Dropbox/Research/Data/Summary_Statisitics/Exercise/SSOE_Model1_BOLTLMM_500K.txt.gz'
  vpa.summary: '~/Dropbox/Research/Data/Summary_Statisitics/Exercise/VPA_Model1_BOLTLMM_500K.txt.gz'
  edu10k.summary: '~/Dropbox/Research/Data/Summary_Statisitics/Education/Lee2018/GWAS_EA.to10K.txt'
  edu.summary: '~/Dropbox/Research/Data/Summary_Statisitics/Education/Lee2018/GWAS_EA_excl23andMe.txt'
  hdl.summary: '~/Dropbox/Research/Data/Summary_Statisitics/Lipids/Willer2013/jointGwasMc_HDL.txt.gz'
  ldl.summary: '~/Dropbox/Research/Data/Summary_Statisitics/Lipids/Willer2013/jointGwasMc_LDL.txt.gz'
  tc.summary: '~/Dropbox/Research/Data/Summary_Statisitics/Lipids/Willer2013/jointGwasMc_TC.txt.gz'
  tg.summary: '~/Dropbox/Research/Data/Summary_Statisitics/Lipids/Willer2013/jointGwasMc_TG.txt.gz'
  sociso.summary: '~/Dropbox/Research/Data/Summary_Statisitics/Social_Isolation/Day2018/MTAG_results.txt.gz'
  insom.summary: '~/Dropbox/Research/Data/Summary_Statisitics/Sleep/Lane2017/Application6818_SleepTrait_data_Lane_RESULTS_DATA_INSOMNIA_SYMPTOMS_all.txt.gz'
  sleep.summary: '~/Dropbox/Research/Data/Summary_Statisitics/Sleep/Lane2017/Application6818_SleepTrait_data_Lane_RESULTS_DATA_SLEEP_DURATION_all.txt.gz'
  agree.summary: ~/Dropbox/Research/Data/Summary_Statisitics/Personality/Lo2016/Agreeablness.txt
  consc.summary: ~/Dropbox/Research/Data/Summary_Statisitics/Personality/Lo2016/Conscientiousness.txt
  extra.summary: ~/Dropbox/Research/Data/Summary_Statisitics/Personality/Lo2016/Extraversion.txt
  neuro.summary: ~/Dropbox/Research/Data/Summary_Statisitics/Personality/Lo2016/Neuroticism.txt
  openn.summary: ~/Dropbox/Research/Data/Summary_Statisitics/Personality/Lo2016/Openness.txt
  ibd.summary: ~/Dropbox/Research/Data/Summary_Statisitics/IBD/deLange2017/ibd_build37_59957_20161107.txt.gz
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(tidyverse)
library(ggplot2)
library(TwoSampleMR)
library(ggman)
library(gridExtra)

##  ---- SNP variance explained ---- ##
# Ftraitoner DS, Mackay TFC. Introduction to Quantitative Genetics, 4th edn. 1996
#  2 x EAF x (1 - EAF) x β^2

snp.r2 <- function(eaf, beta){
  2*eaf*(1 - eaf)*beta^2
}

## p = allele frequency
## n = effective sample size
## z = Z score
z2se <- function(p, n, z){1/sqrt(2*p*(1-p)*(n+z^2))}
```

Read: 
[Tips for Formatting A Lot of GWAS Summary Association Statistics Data](http://huwenboshi.github.io/data%20management/2017/11/23/tips-for-formatting-gwas-summary-stats.html)

[Across-cohort QC analyses of GWAS summary statistics from complex traits](https://doi.org/10.1038/ejhg.2016.106)

[Convert Zscores to Beta and SE](https://www.biostars.org/p/276869/)

### HRC Allele Freq
```{r hrc}
hrc.raw <- read_tsv(params$hrc_file, col_types = cols(CHROM = col_character()))
hrc <- filter(hrc.raw, ID != '.') %>% 
  unite(ChromPos, CHROM, POS, sep = ':', remove = F) %>% 
  unite(MarkerName, ChromPos, REF, ALT, remove = F)
```


## Format GWAS summary statistics for MR analysis 
### Hippocampal Volume 

```{r}
hipv.raw <- read_table2(params$hipv.summary)
hipv.dat <- hipv.raw %>%
  rename(Effect_allele = Allele1, Non_Effect_allele = Allele2, EAF = Freq1, P = `P-value`, SNP = RSNUMBERS) %>% 
  separate(MarkerName, c('CHR', 'POS'), sep = ":") %>% 
  mutate(SE = z2se(EAF, N, Zscore)) %>% 
  mutate(Beta = Zscore*SE) %>% 
  mutate(Effect_allele = toupper(Effect_allele), Non_Effect_allele = toupper(Non_Effect_allele)) %>%
  mutate(r2 = snp.r2(EAF, Beta)) %>%
  filter(nchar(Effect_allele) == 1) %>% 
  filter(Effect_allele %in% c('A', 'C', 'G', 'T')) %>% 
  filter(nchar(Non_Effect_allele) == 1) %>% 
  filter(Non_Effect_allele %in% c('A', 'C', 'G', 'T')) %>% 
  select(SNP, CHR, POS, Effect_allele, Non_Effect_allele, EAF, Zscore, Beta, SE, P, r2, N)
  
write_tsv(hipv.dat, gzfile('~/Dropbox/Research/PostDoc-MSSM/2_MR/2_DerivedData/hipv_GWAS.Processed.gz'))
```

```{r}
## Number of Significant SNPs
mr_hipv.dat <- format_data(filter(hipv.dat, P < 5e-5), type = 'exposure', 
                           phenotype_col = 'hipv',
                           snp_col = 'SNP', 
                           beta_col = "Beta",
                           se_col = "SE",
                           eaf_col = "EAF",
                           effect_allele_col = "Effect_allele",
                           other_allele_col = "Non_Effect_allele",
                           pval_col = "P")
hipv_ld <- clump_data(mr_hipv.dat, clump_r2 = 0.1, clump_kb = 250)

clump.n <- hipv_ld %>% 
  mutate(p.cut = cut(pval.exposure, breaks = 
                       c(Inf, 0.05, 5e-5, 5e-6, 5e-7, 5e-8, -Inf), lablels = F)) %>%
  left_join(select(hipv.dat, SNP, r2)) %>% 
  select(SNP, effect_allele.exposure, other_allele.exposure, eaf.exposure, beta.exposure, se.exposure, p.cut, r2) %>% 
  group_by(p.cut) %>% summarise(n = n(), r2 = round(sum(r2, na.rm = T), 4))

all.n <- hipv.dat %>% 
  mutate(p.cut = cut(P, breaks = 
                       c(Inf, 0.05, 5e-5, 5e-6, 5e-7, 5e-8, -Inf), lablels = F)) %>% 
  group_by(p.cut) %>% summarise(n = n(), r2 = round(sum(r2, na.rm = T), 4))

tab.hipv <- full_join(all.n, clump.n, by = 'p.cut', suffix = c('.all', '.clump'))
```

```{r alc_cons gwas}

p.hipv <- ggman(filter(hipv.dat, P < 0.05), snp = 'SNP', chrom = 'CHR', bp = 'POS', pvalue = 'P', ymin = 0, ymax = 30, 
                title = "hipv - Lambert 2013") + 
  theme_classic() + 
  geom_hline(yintercept = -log10(5e-5), colour = 'blue', size = 0.25) + 
  geom_hline(yintercept = -log10(5e-6), colour = 'red', size = 0.25, linetype = 2)
p.hipv <- ggmanHighlight(p.hipv, highlight = filter(hipv_ld, pval.exposure < 5e-6)$SNP, shape = 18)

out.hipv <- grid.arrange(p.hipv,  tableGrob(tab.hipv, theme = ttheme_minimal(base_size = 8)), 
                         ncol = 2, widths = 2:1, layout_matrix = rbind(c(1,2), c(1, NA)))

ggsave('~/Dropbox/Research/PostDoc-MSSM/2_MR/4_Output/hipv.png', device = 'png', units = 'in', width = 10, height = 5, plot = out.hipv)
```

### Late Onset Alzheimer's Disease
[Lambert et al Nature Genet 2013](https://www.nature.com/articles/ng.2802): The International Genomics of Alzheimer’s Project (IGAP) is a meta-analysis of 4 previously published GWAS datasets: the European Alzheimer’s Disease Imitative (EADI), the Alzheimer Disease Genetics Consortium (ADGC), Cohorts for Heart and Aging Research in Genomic Epidemiology (CHARGE), and Genetic and Environmental Risk in AD (GERAD) and includes a sample of 17,008 LOAD cases and 37,154 cognitively normal elder controls. Participants in IGAP were of European ancestry, the average age was 71 and 58.4% of participants were women.

```{r data_wrangling_outcomes, echo=F, cache=TRUE}
igap.dat.raw <- read_tsv(params$load.summary)
aaos.dat.raw <- read_tsv(params$aaos.summary)
```

```{r}
igap.dat <- igap.dat.raw %>%
  left_join(select(aaos.dat.raw, SNP, Freq1), by = 'SNP') %>%
  rename(EAF = Freq1) %>%
  filter(nchar(Effect_allele) == 1) %>% 
  filter(nchar(Non_Effect_allele) == 1) %>% 
  mutate(r2 = snp.r2(EAF, Beta), N = 54162) %>%
  select(SNP, CHR, POS, Effect_allele, Non_Effect_allele, EAF, Beta, SE, P, r2, N)

write_tsv(igap.dat, gzfile('~/Dropbox/Research/PostDoc-MSSM/2_MR/2_DerivedData/load_GWAS.Processed.gz'))
```

```{r}
## Number of Significant SNPs
mr_load.dat <- format_data(filter(igap.dat, P < 5e-5), type = 'exposure', 
                               phenotype_col = 'load',
                               snp_col = 'SNP', 
                               beta_col = "Beta",
                               se_col = "SE",
                               eaf_col = "EAF",
                               effect_allele_col = "Effect_allele",
                               other_allele_col = "Non_Effect_allele",
                               pval_col = "P")
load_ld <- clump_data(mr_load.dat, clump_r2 = 0.1, clump_kb = 250)

clump.n <- load_ld %>% 
  mutate(p.cut = cut(pval.exposure, breaks = 
                       c(Inf, 0.05, 5e-5, 5e-6, 5e-7, 5e-8, -Inf), lablels = F)) %>%
  left_join(select(igap.dat, SNP, r2)) %>% 
  select(SNP, effect_allele.exposure, other_allele.exposure, eaf.exposure, beta.exposure, se.exposure, p.cut, r2) %>% 
  group_by(p.cut) %>% summarise(n = n(), r2 = round(sum(r2, na.rm = T), 4))

all.n <- igap.dat %>% 
  mutate(p.cut = cut(P, breaks = 
                       c(Inf, 0.05, 5e-5, 5e-6, 5e-7, 5e-8, -Inf), lablels = F)) %>% 
  group_by(p.cut) %>% summarise(n = n(), r2 = round(sum(r2, na.rm = T), 4))

tab.load <- full_join(all.n, clump.n, by = 'p.cut', suffix = c('.all', '.clump'))
```

```{r alc_cons gwas}

p.load <- ggman(filter(igap.dat, P < 0.05), snp = 'SNP', chrom = 'CHR', bp = 'POS', pvalue = 'P', ymin = 0, ymax = 60, 
                title = "LOAD - Lambert 2013") + 
  theme_classic() + 
  geom_hline(yintercept = -log10(5e-5), colour = 'blue', size = 0.25) + 
  geom_hline(yintercept = -log10(5e-6), colour = 'red', size = 0.25, linetype = 2)
p.load <- ggmanHighlight(p.load, highlight = filter(load_ld, pval.exposure < 5e-6)$SNP, shape = 18)

out.load <- grid.arrange(p.load,  
                         tableGrob(tab.load, theme = ttheme_minimal(base_size = 8)), 
                         ncol = 2, widths = 2:1, layout_matrix = rbind(c(1,2), c(1, NA)))

ggsave('~/Dropbox/Research/PostDoc-MSSM/2_MR/4_Output/load.png', device = 'png', units = 'in', width = 10, height = 5, plot = out.load)
```

### Alzheimer's Age of Onset Surivial
[Huang et al 2017](https://www.nature.com/articles/nn.4587): A GWAS of age of onset in LOAD was conducted in 14,406 AD case samples and 25,849 control samples from the IGAP using Cox proportional hazard regressions. Participants were of European ancestry, in cases the the average AAO was 74.8 and 61.7% were women, in controls the average AAE was 79.0 and 59.6% were women.
```{r}
aaos.dat.raw <- read_tsv(params$aaos.summary)
aaos.dat <- aaos.dat.raw %>%
  rename(POS = BP, Effect_allele = Allele1, Non_Effect_allele = Allele2, EAF = Freq1, SE = StdErr, Beta = Effect) %>%
  arrange(CHR, POS) %>%
  mutate(Effect_allele = toupper(Effect_allele), Non_Effect_allele = toupper(Non_Effect_allele)) %>%
  filter(nchar(Effect_allele) == 1) %>% filter(Effect_allele %in% c('A', 'C', 'G', 'T')) %>% 
  filter(nchar(Non_Effect_allele) == 1) %>% filter(Non_Effect_allele %in% c('A', 'C', 'G', 'T')) %>% 
  mutate(r2 = snp.r2(EAF, Beta), N = 40255) %>% 
  select(SNP, CHR, POS, Effect_allele, Non_Effect_allele, EAF, Beta, SE, P, r2, N)

write_tsv(aaos.dat, gzfile('~/Dropbox/Research/PostDoc-MSSM/2_MR/2_DerivedData/aaos_GWAS.Processed.gz'))
```

```{r}
## Number of Significant SNPs
mr_aaos.dat <- format_data(filter(aaos.dat, P < 5e-5), type = 'exposure', 
                               phenotype_col = 'aaos',
                               snp_col = 'SNP', 
                               beta_col = "Beta",
                               se_col = "SE",
                               eaf_col = "EAF",
                               effect_allele_col = "Effect_allele",
                               other_allele_col = "Non_Effect_allele",
                               pval_col = "P")
aaos_ld <- clump_data(mr_aaos.dat, clump_r2 = 0.1, clump_kb = 250)

clump.n <- aaos_ld %>% 
  mutate(p.cut = cut(pval.exposure, breaks = 
                       c(Inf, 0.05, 5e-5, 5e-6, 5e-7, 5e-8, -Inf), lablels = F)) %>%
  left_join(select(aaos.dat, SNP, r2)) %>% 
  select(SNP, effect_allele.exposure, other_allele.exposure, eaf.exposure, beta.exposure, se.exposure, p.cut, r2) %>% 
  group_by(p.cut) %>% summarise(n = n(), r2 = round(sum(r2, na.rm = T), 4))

all.n <- aaos.dat %>% 
  mutate(p.cut = cut(P, breaks = 
                       c(Inf, 0.05, 5e-5, 5e-6, 5e-7, 5e-8, -Inf), lablels = F)) %>% 
  group_by(p.cut) %>% summarise(n = n(), r2 = round(sum(r2, na.rm = T), 4))

tab.aaos <- full_join(all.n, clump.n, by = 'p.cut', suffix = c('.all', '.clump'))
```

```{r alc_cons gwas}

p.aaos <- ggman(filter(aaos.dat, P < 0.05), snp = 'SNP', chrom = 'CHR', bp = 'POS', pvalue = 'P', ymin = 0, ymax = 60, 
                title = "Alzheimer's Age of Onset - Huang 2017") + 
  theme_classic() + 
  geom_hline(yintercept = -log10(5e-5), colour = 'blue', size = 0.25) + 
  geom_hline(yintercept = -log10(5e-6), colour = 'red', size = 0.25, linetype = 2) 
  
p.aaos <- ggmanHighlight(p.aaos, highlight = filter(aaos_ld, pval.exposure < 5e-6)$SNP, shape = 18)

out.aaos <- grid.arrange(p.aaos,  tableGrob(tab.aaos, theme = ttheme_minimal(base_size = 8)), 
                         ncol = 2, widths = 2:1, layout_matrix = rbind(c(1,2), c(1, NA)))

ggsave('~/Dropbox/Research/PostDoc-MSSM/2_MR/4_Output/aaos.png', device = 'png', units = 'in', width = 10, height = 5, plot = out.aaos)
```

## CSF AB42, tau & ptau
[Deming et al Acta Neuropathologica 2017](https://link.springer.com/article/10.1007%2Fs00401-017-1685-y): A GWAS of CSF AB42, ptau and tau levels (pg/mL) was conducted in 3,146 participants. Participants were of Eurpean ancestry.

### AB42
```{r}
## ---- AB42 ---- ##
#  rename and reorder columns
ab42.raw <- read_table2(params$ab42.summary)
ab42 <- ab42.raw %>%
  left_join(select(hrc, ID, REF, ALT, AF), by = c('SNP' = 'ID')) %>%
  mutate(A2 = ifelse(A1 == REF, ALT, REF)) %>% 
  mutate(EAF = ifelse(A1 == ALT, AF, 1-AF)) %>%
  rename(Beta = BETA, POS = BP, Effect_allele = A1, Non_Effect_allele = A2, N = NMISS) %>%
  filter(nchar(Effect_allele) == 1) %>%
  filter(nchar(Non_Effect_allele) == 1) %>% 
  mutate(r2 = snp.r2(EAF, Beta)) %>% 
  select(SNP, CHR, POS, Effect_allele, Non_Effect_allele, EAF, Beta, SE, P, r2, N)

write_tsv(ab42, gzfile('~/Dropbox/Research/PostDoc-MSSM/2_MR/2_DerivedData/ab42_GWAS.Processed.gz'))
```

```{r}
## Number of Significant SNPs
mr_ab42.dat <- format_data(filter(ab42, P < 5e-5), type = 'exposure', 
                               phenotype_col = 'ab42',
                               snp_col = 'SNP', 
                               beta_col = "Beta",
                               se_col = "SE",
                               eaf_col = "EAF",
                               effect_allele_col = "Effect_allele",
                               other_allele_col = "Non_Effect_allele",
                               pval_col = "P")
ab42_ld <- clump_data(mr_ab42.dat, clump_r2 = 0.1, clump_kb = 250)

clump.n <- ab42_ld %>% 
  mutate(p.cut = cut(pval.exposure, breaks = 
                       c(Inf, 0.05, 5e-5, 5e-6, 5e-7, 5e-8, -Inf), lablels = F)) %>%
  left_join(select(ab42, SNP, r2)) %>% 
  select(SNP, effect_allele.exposure, other_allele.exposure, eaf.exposure, beta.exposure, se.exposure, p.cut, r2) %>% 
  group_by(p.cut) %>% summarise(n = n(), r2 = round(sum(r2, na.rm = T), 4))

all.n <- ab42 %>% 
  mutate(p.cut = cut(P, breaks = 
                       c(Inf, 0.05, 5e-5, 5e-6, 5e-7, 5e-8, -Inf), lablels = F)) %>% 
  group_by(p.cut) %>% summarise(n = n(), r2 = round(sum(r2, na.rm = T), 4))

tab.ab42 <- full_join(all.n, clump.n, by = 'p.cut', suffix = c('.all', '.clump'))
```

```{r alc_cons gwas}

p.ab42 <- ggman(filter(ab42, P < 0.05), snp = 'SNP', chrom = 'CHR', bp = 'POS', pvalue = 'P', ymin = 0, ymax = 20, 
                title = "CSF AB42 - Deming 2017") + 
  theme_classic() + 
  geom_hline(yintercept = -log10(5e-5), colour = 'blue', size = 0.25) + 
  geom_hline(yintercept = -log10(5e-6), colour = 'red', size = 0.25, linetype = 2)
p.ab42 <- ggmanHighlight(p.ab42, highlight = filter(ab42_ld, pval.exposure < 5e-6)$SNP, shape = 18)

out.ab42 <- grid.arrange(p.ab42,  tableGrob(tab.ab42, theme = ttheme_minimal(base_size = 8)), 
                         ncol = 2, widths = 2:1, layout_matrix = rbind(c(1,2), c(1, NA)))

ggsave('~/Dropbox/Research/PostDoc-MSSM/2_MR/4_Output/ab42.png', device = 'png', units = 'in', width = 10, height = 5, plot = out.ab42)
```

### Ptau
```{r}
## ---- ptau ---- ##
#  rename and reorder columns
ptau.raw <- read_table2(params$ptau.summary)
ptau <- ptau.raw %>%
  left_join(select(hrc, ID, REF, ALT, AF), by = c('SNP' = 'ID')) %>%
  mutate(A2 = ifelse(A1 == REF, ALT, REF)) %>% 
  mutate(EAF = ifelse(A1 == ALT, AF, 1-AF)) %>%
  rename(Beta = BETA, POS = BP, Effect_allele = A1, Non_Effect_allele = A2, N = NMISS) %>%
  filter(nchar(Effect_allele) == 1) %>%
  filter(nchar(Non_Effect_allele) == 1) %>% 
  mutate(r2 = snp.r2(EAF, Beta)) %>% 
  select(SNP, CHR, POS, Effect_allele, Non_Effect_allele, EAF, Beta, SE, P, r2, N)

write_tsv(ptau, gzfile('~/Dropbox/Research/PostDoc-MSSM/2_MR/2_DerivedData/ptau_GWAS.Processed.gz'))
```

```{r}
## Number of Significant SNPs
mr_ptau.dat <- format_data(filter(ptau, P < 5e-5), type = 'exposure', 
                               phenotype_col = 'ptau',
                               snp_col = 'SNP', 
                               beta_col = "Beta",
                               se_col = "SE",
                               eaf_col = "EAF",
                               effect_allele_col = "Effect_allele",
                               other_allele_col = "Non_Effect_allele",
                               pval_col = "P")
ptau_ld <- clump_data(mr_ptau.dat, clump_r2 = 0.1, clump_kb = 250)

clump.n <- ptau_ld %>% 
  mutate(p.cut = cut(pval.exposure, breaks = 
                       c(Inf, 0.05, 5e-5, 5e-6, 5e-7, 5e-8, -Inf), lablels = F)) %>%
  left_join(select(ptau, SNP, r2)) %>% 
  select(SNP, effect_allele.exposure, other_allele.exposure, eaf.exposure, beta.exposure, se.exposure, p.cut, r2) %>% 
  group_by(p.cut) %>% summarise(n = n(), r2 = round(sum(r2, na.rm = T), 4))

all.n <- ptau %>% 
  mutate(p.cut = cut(P, breaks = 
                       c(Inf, 0.05, 5e-5, 5e-6, 5e-7, 5e-8, -Inf), lablels = F)) %>% 
  group_by(p.cut) %>% summarise(n = n(), r2 = round(sum(r2, na.rm = T), 4))

tab.ptau <- full_join(all.n, clump.n, by = 'p.cut', suffix = c('.all', '.clump'))
```

```{r alc_cons gwas}

p.ptau <- ggman(filter(ptau, P < 0.05), snp = 'SNP', chrom = 'CHR', bp = 'POS', pvalue = 'P', ymin = 0, ymax = 20, 
                title = "CSF Ptau - Deming 2017") + 
  theme_classic() + 
  geom_hline(yintercept = -log10(5e-5), colour = 'blue', size = 0.25) + 
  geom_hline(yintercept = -log10(5e-6), colour = 'red', size = 0.25, linetype = 2)
p.ptau <- ggmanHighlight(p.ptau, highlight = filter(ptau_ld, pval.exposure < 5e-6)$SNP, shape = 18)

out.ptau <- grid.arrange(p.ptau,  tableGrob(tab.ptau, theme = ttheme_minimal(base_size = 8)), 
                         ncol = 2, widths = 2:1, layout_matrix = rbind(c(1,2), c(1, NA)))

ggsave('~/Dropbox/Research/PostDoc-MSSM/2_MR/4_Output/ptau.png', device = 'png', units = 'in', width = 10, height = 5, plot = out.ptau)
```

### Tau
```{r}
## ---- tau ---- ##
#  rename and reorder columns
tau.raw <- read_table2(params$tau.summary)
tau <- tau.raw %>%
  left_join(select(hrc, ID, REF, ALT, AF), by = c('SNP' = 'ID')) %>%
  mutate(A2 = ifelse(A1 == REF, ALT, REF)) %>% 
  mutate(EAF = ifelse(A1 == ALT, AF, 1-AF)) %>%
  rename(Beta = BETA, POS = BP, Effect_allele = A1, Non_Effect_allele = A2, N = NMISS) %>%
  filter(nchar(Effect_allele) == 1) %>%
  filter(nchar(Non_Effect_allele) == 1) %>% 
  mutate(r2 = snp.r2(EAF, Beta)) %>% 
  select(SNP, CHR, POS, Effect_allele, Non_Effect_allele, EAF, Beta, SE, P, r2, N)

write_tsv(tau.dat, gzfile('~/Dropbox/Research/PostDoc-MSSM/2_MR/2_DerivedData/tau_GWAS.Processed.gz'))
```

```{r}
## Number of Significant SNPs
mr_tau.dat <- format_data(filter(tau, P < 5e-5), type = 'exposure', 
                               phenotype_col = 'tau',
                               snp_col = 'SNP', 
                               beta_col = "Beta",
                               se_col = "SE",
                               eaf_col = "EAF",
                               effect_allele_col = "Effect_allele",
                               other_allele_col = "Non_Effect_allele",
                               pval_col = "P")
tau_ld <- clump_data(mr_tau.dat, clump_r2 = 0.1, clump_kb = 250)

clump.n <- tau_ld %>% 
  mutate(p.cut = cut(pval.exposure, breaks = 
                       c(Inf, 0.05, 5e-5, 5e-6, 5e-7, 5e-8, -Inf), lablels = F)) %>%
  left_join(select(tau, SNP, r2)) %>% 
  select(SNP, effect_allele.exposure, other_allele.exposure, eaf.exposure, beta.exposure, se.exposure, p.cut, r2) %>% 
  group_by(p.cut) %>% summarise(n = n(), r2 = round(sum(r2, na.rm = T), 4))

all.n <- tau %>% 
  mutate(p.cut = cut(P, breaks = 
                       c(Inf, 0.05, 5e-5, 5e-6, 5e-7, 5e-8, -Inf), lablels = F)) %>% 
  group_by(p.cut) %>% summarise(n = n(), r2 = round(sum(r2, na.rm = T), 4))

tab.tau <- full_join(all.n, clump.n, by = 'p.cut', suffix = c('.all', '.clump'))
```

```{r alc_cons gwas}

p.tau <- ggman(filter(tau, P < 0.05), snp = 'SNP', chrom = 'CHR', bp = 'POS', pvalue = 'P', ymin = 0, ymax = 30, 
                title = "CSF Tau - Deming 2017") + 
  theme_classic() + 
  geom_hline(yintercept = -log10(5e-5), colour = 'blue', size = 0.25) + 
  geom_hline(yintercept = -log10(5e-6), colour = 'red', size = 0.25, linetype = 2)
p.tau <- ggmanHighlight(p.tau, highlight = filter(tau_ld, pval.exposure < 5e-6)$SNP, shape = 18)

out.tau <- grid.arrange(p.tau,  tableGrob(tab.tau, theme = ttheme_minimal(base_size = 8)), 
                         ncol = 2, widths = 2:1, layout_matrix = rbind(c(1,2), c(1, NA)))

ggsave('~/Dropbox/Research/PostDoc-MSSM/2_MR/4_Output/tau.png', device = 'png', units = 'in', width = 10, height = 5, plot = out.tau)
```

## Alcohol Consumption
[Clarke et al Mol Psychiatr 2017](https://www.nature.com/articles/mp2017153): SNPs associated with alcohol consumption were selected from a GWAS of alcohol consumption (drinks per week adjusted for sex, age and weight) performed in 112,117 individuals from the UK Biobank. Participants were of European ancestry, mean age was 59.6 and 52.7% of the cohort was female.
```{r data_wrangling_exposure, echo=F, cache=TRUE}
## Trait Code: alcc
alc_cons.raw <- read_table2(params$alc_cons.summary)
alc_cons <- alc_cons.raw %>% 
  filter(nchar(A1) == 1) %>% 
  filter(nchar(A2) == 1) %>% 
  rename(Effect_allele = A1, Non_Effect_allele = A2, Beta = BETA) %>% 
  mutate(r2 = snp.r2(EAF, Beta)) %>%
  select(SNP, CHR, POS, Effect_allele, Non_Effect_allele, EAF, Beta, SE, P, r2, N)

write_tsv(alc_cons, gzfile('~/Dropbox/Research/PostDoc-MSSM/2_MR/2_DerivedData/alcc_GWAS.Processed.gz'))

```

```{r}
## Number of Significant SNPs
mr_trait.dat <- format_data(filter(alc_cons, P < 5e-5), type = 'exposure', 
                               phenotype_col = 'trait',
                               snp_col = 'SNP', 
                               beta_col = "Beta",
                               se_col = "SE",
                               eaf_col = "EAF",
                               effect_allele_col = "Effect_allele",
                               other_allele_col = "Non_Effect_allele",
                               pval_col = "P")
mr_trait.dat_ld <- clump_data(mr_trait.dat, clump_r2 = 0.1, clump_kb = 250)

clump.n <- mr_trait.dat_ld %>% 
  mutate(p.cut = cut(pval.exposure, breaks = c(Inf, 0.05, 5e-5, 5e-6, 5e-7, 5e-8, -Inf), lablels = F)) %>% 
  left_join(select(alc_cons, SNP, r2)) %>% 
  select(SNP, effect_allele.exposure, other_allele.exposure, eaf.exposure, beta.exposure, se.exposure, p.cut, r2) %>% 
  group_by(p.cut) %>% summarise(n = n(), r2 = sum(r2))
all.n <- alc_cons %>% 
  mutate(p.cut = cut(P, breaks = c(Inf, 0.05, 5e-5, 5e-6, 5e-7, 5e-8, -Inf), lablels = F)) %>% 
  group_by(p.cut) %>% summarise(n = n(), r2 = sum(r2))

full_join(all.n, clump.n, by = 'p.cut', suffix = c('.all', '.clump'))

```

```{r alc_cons gwas}

ggman(filter(alc_cons, P < 0.05), snp = 'SNP', chrom = 'CHR', bp = 'POS', pvalue = 'P', ymin = 0, ymax = 20, title = "Alcohol Consumption - Clarke 2017") + theme_classic() + geom_hline(yintercept = -log10(5e-5), colour = 'blue', size = 0.25) 
ggsave('~/Dropbox/Research/PostDoc-MSSM/2_MR/4_Output/alc_cons.png', device = 'png', units = 'in', width = 9, height = 5)
```

## Alcohol Dependence
[Walters et al BioRxiv 2018](https://www.biorxiv.org/content/early/2018/03/10/257311): SNPs associated with alcohol dependnece were selected from a GWAS of DSM-IV diagnosed alcohol dependnece perfomred in 46,568 individules (11,569 cases and 34,999 controls) of European ancestry.  

```{r data_wrangling_exposure, echo=F, cache=TRUE}
alc_dep.raw <- read_table2(params$alc_dep.summary)
alc_dep <- alc_dep.raw %>% 
  rename(Effect_allele = A1, Non_Effect_allele = A2, POS = BP, N = Neff) %>% 
  arrange(CHR, POS) %>% 
  filter(nchar(Effect_allele) == 1) %>% 
  filter(nchar(Non_Effect_allele) == 1) %>% 
  mutate(Beta = log(OR)) %>%
  mutate(EAF = ((Nca*FRQ_A_8485) + (Nco*FRQ_U_20272)) / (Nca + Nco)) %>%
  mutate(r2 = snp.r2(EAF, Beta)) %>%
  select(SNP, CHR, POS, Effect_allele, Non_Effect_allele, EAF, Beta, SE, P, r2, N)

write_tsv(alc_dep, gzfile('~/Dropbox/Research/PostDoc-MSSM/2_MR/2_DerivedData/alcd_GWAS.Processed.gz'))

```

```{r}
## Number of Significant SNPs
alc_dep_ld <- format_data(filter(alc_dep, P < 5e-5), type = 'exposure', 
                               phenotype_col = 'trait',
                               snp_col = 'SNP', 
                               beta_col = "Beta",
                               se_col = "SE",
                               eaf_col = "EAF",
                               effect_allele_col = "Effect_allele",
                               other_allele_col = "Non_Effect_allele",
                               pval_col = "P")
alc_dep_ld <- clump_data(alc_dep_ld, clump_r2 = 0.1, clump_kb = 250)

clump.n <- alc_dep_ld %>% 
  mutate(p.cut = cut(pval.exposure, breaks = c(Inf, 0.05, 5e-5, 5e-6, 5e-7, 5e-8, -Inf), lablels = F)) %>% 
  left_join(select(alc_dep, SNP, r2)) %>% 
  select(SNP, effect_allele.exposure, other_allele.exposure, eaf.exposure, beta.exposure, se.exposure, p.cut, r2) %>% 
  group_by(p.cut) %>% summarise(n = n(), r2 = sum(r2))
all.n <- alc_dep %>% 
  mutate(p.cut = cut(P, breaks = c(Inf, 0.05, 5e-5, 5e-6, 5e-7, 5e-8, -Inf), lablels = F)) %>% 
  group_by(p.cut) %>% summarise(n = n(), r2 = sum(r2))

full_join(all.n, clump.n, by = 'p.cut', suffix = c('.all', '.clump'))
```

```{r alc_cons gwas}

ggman(filter(alc_dep, P < 0.05), snp = 'SNP', chrom = 'CHR', bp = 'POS', pvalue = 'P', ymin = 0, ymax = 15, title = "Alcohol Dependence - Walters 2018") + theme_classic() + geom_hline(yintercept = -log10(5e-5), colour = 'blue', size = 0.25) 
ggsave('~/Dropbox/Research/PostDoc-MSSM/2_MR/4_Output/alc_dep.png', device = 'png', units = 'in', width = 9, height = 5)
```

## AUDIT 
[Sanchez-Roige et al BioRxiv 2018](https://www.biorxiv.org/content/early/2018/08/02/275917): A GWAS of the Alcohol Use Disorder Identification Tes (AUDIT) was conducted in the British white individules in the UK Biobank. The GWAS meta-analysis with 23andMe, identified 10 genome-wide significant loci with SNP-heritability of 12%.   

### AUDIT Total Score
```{r data_wrangling_exposure, echo=F, cache=TRUE}
audit.raw <- read_table2(params$audit.summary)
audit <- audit.raw %>% 
  rename(Effect_allele = a_1, Non_Effect_allele = a_0, SNP = rsid) %>% 
  filter(nchar(Effect_allele) == 1) %>% 
  filter(nchar(Non_Effect_allele) == 1) %>% 
  left_join(select(hrc, ID, POS, REF, ALT, AF), by = c('SNP' = 'ID')) %>%
  mutate(EAF = ifelse(Effect_allele == ALT, AF, 1-AF)) 

audit_t <- audit %>% 
  select(SNP, chr, POS, Effect_allele, Non_Effect_allele, EAF, beta_T, se_T, p_T, N) %>% 
  rename(CHR = chr, Beta = beta_T, SE = se_T, P = p_T) %>%
  mutate(r2 = snp.r2(EAF, Beta))
  
audit_c <- audit %>% 
  select(SNP, chr, POS, Effect_allele, Non_Effect_allele, EAF, beta_C, se_C, p_C, N) %>% 
  rename(CHR = chr, Beta = beta_C, SE = se_C, P = p_C) %>%
  mutate(r2 = snp.r2(EAF, Beta))

audit_p <- audit %>% 
  select(SNP, chr, POS, Effect_allele, Non_Effect_allele, EAF, beta_P, se_P, p_P, N) %>% 
  rename(CHR = chr, Beta = beta_P, SE = se_P, P = p_P) %>%
  mutate(r2 = snp.r2(EAF, Beta))

write_tsv(audit_t, gzfile('~/Dropbox/Research/PostDoc-MSSM/2_MR/2_DerivedData/audit_GWAS.Processed.gz'))

```

```{r}
## Number of Significant SNPs
audit_ld <- format_data(filter(audit_t, P < 5e-5), type = 'exposure', 
                               phenotype_col = 'trait',
                               snp_col = 'SNP', 
                               beta_col = "Beta",
                               se_col = "SE",
                               eaf_col = "EAF",
                               effect_allele_col = "Effect_allele",
                               other_allele_col = "Non_Effect_allele",
                               pval_col = "P")
audit_ld <- clump_data(audit_ld, clump_r2 = 0.1, clump_kb = 250)

clump.n <- audit_ld %>% 
  mutate(p.cut = cut(pval.exposure, breaks = c(Inf, 0.05, 5e-5, 5e-6, 5e-7, 5e-8, -Inf), lablels = F)) %>% 
  left_join(select(audit_t, SNP, r2)) %>% 
  select(SNP, effect_allele.exposure, other_allele.exposure, eaf.exposure, beta.exposure, se.exposure, p.cut, r2) %>% 
  group_by(p.cut) %>% summarise(n = n(), r2 = sum(r2, na.rm = T))
all.n <- audit_t %>% 
  mutate(p.cut = cut(P, breaks = c(Inf, 0.05, 5e-5, 5e-6, 5e-7, 5e-8, -Inf), lablels = F)) %>% 
  group_by(p.cut) %>% summarise(n = n(), r2 = sum(r2, na.rm = T))

full_join(all.n, clump.n, by = 'p.cut', suffix = c('.all', '.clump'))
```

```{r alc_cons gwas}

ggman(filter(audit_t, P < 0.05), snp = 'SNP', chrom = 'CHR', bp = 'POS', pvalue = 'P', ymin = 0, ymax = 20, title = "AUDIT Total - Sanchez-Roige 2018") + theme_classic() + geom_hline(yintercept = -log10(5e-5), colour = 'blue', size = 0.25) + geom_hline(yintercept = -log10(5e-6), colour = 'red', size = 0.25, linetype = 2) 
ggsave('~/Dropbox/Research/PostDoc-MSSM/2_MR/4_Output/audit.png', device = 'png', units = 'in', width = 9, height = 5)
```

### AUDIT Consumption
```{r data_wrangling_exposure, echo=F, cache=TRUE}
audit.raw <- read_table2(params$audit.summary)
audit <- audit.raw %>% 
  rename(Effect_allele = a_1, Non_Effect_allele = a_0, SNP = rsid) %>% 
  filter(nchar(Effect_allele) == 1) %>% 
  filter(nchar(Non_Effect_allele) == 1) %>% 
  left_join(select(hrc, ID, POS, REF, ALT, AF), by = c('SNP' = 'ID')) %>%
  mutate(EAF = ifelse(Effect_allele == ALT, AF, 1-AF)) 

audit_c <- audit %>% 
  select(SNP, chr, POS, Effect_allele, Non_Effect_allele, EAF, beta_C, se_C, p_C, N) %>% 
  rename(CHR = chr, Beta = beta_C, SE = se_C, P = p_C) %>%
  mutate(r2 = snp.r2(EAF, Beta))

write_tsv(audit_c, gzfile('~/Dropbox/Research/PostDoc-MSSM/2_MR/2_DerivedData/audit_c_GWAS.Processed.gz'))

```

### AUDIT Problem Drinking
```{r data_wrangling_exposure, echo=F, cache=TRUE}
audit.raw <- read_table2(params$audit.summary)
audit <- audit.raw %>% 
  rename(Effect_allele = a_1, Non_Effect_allele = a_0, SNP = rsid) %>% 
  filter(nchar(Effect_allele) == 1) %>% 
  filter(nchar(Non_Effect_allele) == 1) %>% 
  left_join(select(hrc, ID, POS, REF, ALT, AF), by = c('SNP' = 'ID')) %>%
  mutate(EAF = ifelse(Effect_allele == ALT, AF, 1-AF)) 

audit_p <- audit %>% 
  select(SNP, chr, POS, Effect_allele, Non_Effect_allele, EAF, beta_P, se_P, p_P, N) %>% 
  rename(CHR = chr, Beta = beta_P, SE = se_P, P = p_P) %>%
  mutate(r2 = snp.r2(EAF, Beta))

write_tsv(audit_p, gzfile('~/Dropbox/Research/PostDoc-MSSM/2_MR/2_DerivedData/audit_p_GWAS.Processed.gz'))

```

## Body Mass Index
[Yengo et al BioRxiv 2018](https://www.biorxiv.org/content/early/2018/03/22/274654): Meta-analysis of Giant Consortium GWAS of BMI (n = 234,069) and a new GWAS of BMI conducted in the UK Biobank (n = 456,426) for a total of 681,275 participants of European ancestry. 

```{r data_wrangling_exposure, echo=F, cache=TRUE}
bmi.raw <- read_table2(params$bmi.summary)
bmi <- bmi.raw %>% 
  rename(Effect_allele = Tested_Allele, Non_Effect_allele = Other_Allele, Beta = BETA, EAF = Freq_Tested_Allele_in_HRS) %>% 
  filter(nchar(Effect_allele) == 1) %>% 
  filter(nchar(Non_Effect_allele) == 1) %>% 
  arrange(CHR, POS) %>% 
  mutate(r2 = snp.r2(EAF, Beta)) %>%
  select(SNP, CHR, POS, Effect_allele, Non_Effect_allele, EAF, Beta, SE, P, r2, N)

write_tsv(bmi, gzfile('~/Dropbox/Research/PostDoc-MSSM/2_MR/2_DerivedData/bmi_GWAS.Processed.gz'))
```

```{r}
## Clump snps
system('plink --bfile /Users/sheaandrews/Documents/0_Data/1000genomes/EUR/EUR_All_Chr --clump ~/Dropbox/Research/Data/Summary_Statisitics/BMI/Yengo2018/BMI_gwas.assoc --clump-r2 0.1 --clump-kb 250 --clump-p1 1 --clump-p2 1 --out ~/Dropbox/Research/Data/Summary_Statisitics/BMI/Yengo2018/bmi_clump')

bmi_ld <- read_table2('~/Dropbox/Research/Data/Summary_Statisitics/BMI/Yengo2018/bmi_clump.clumped') %>% 
  filter(!is.na(CHR))

clump.n <- bmi_ld %>% 
  mutate(p.cut = cut(P, breaks = c(Inf, 0.05, 5e-5, 5e-6, 5e-7, 5e-8, -Inf), lablels = F)) %>% 
  left_join(select(bmi, SNP, r2)) %>% 
  group_by(p.cut) %>% summarise(n = n(), r2 = round(sum(r2), 4))

all.n <- bmi %>% 
  mutate(p.cut = cut(P, breaks = c(Inf, 0.05, 5e-5, 5e-6, 5e-7, 5e-8, -Inf), lablels = F)) %>% 
  group_by(p.cut) %>% summarise(n = n(), r2 = round(sum(r2), 4))

tab.bmi <- full_join(all.n, clump.n, by = 'p.cut', suffix = c('.all', '.clump'))
```

```{r alc_cons gwas}

p.bim <- ggman(filter(bmi, P < 0.05), snp = 'SNP', chrom = 'CHR', bp = 'POS', pvalue = 'P', ymin = 0, ymax = 60, 
      title = "BMI - Yengo 2018") + 
  theme_classic() + 
  geom_hline(yintercept = -log10(5e-5), colour = 'blue', size = 0.25) + 
  geom_hline(yintercept = -log10(5e-6), colour = 'red', size = 0.25, linetype = 2)

out.bmi <- grid.arrange(p.bim,  tableGrob(tab.bmi, theme = ttheme_minimal(base_size = 8)), 
             ncol = 2, widths = 2:1, layout_matrix = rbind(c(1,2), c(1, NA)))

ggsave('~/Dropbox/Research/PostDoc-MSSM/2_MR/4_Output/bmi.png', device = 'png', units = 'in', width = 10, height = 5, plot = out.bmi)
```


## Major Depressive Disorder
[Wray et al Nature Genet 2018](https://www.nature.com/articles/s41588-018-0090-3): The Psychiatric Genomics Consortium (PGC) conducted a GWAS meta-analysis across seven cohorts (PGC, deCODE, Generation Scotland, GERA, iPSYCH, UK Biobank, and 23andMe) containing 130,664 MDD cases and 330,470 controls. Participants were of European Ancestry. 

```{r data_wrangling_exposure, echo=F, cache=TRUE}
mdd.raw <- read_table2(params$mdd.summary)
mdd <- mdd.raw %>% 
  mutate(EAF = ((135458*FRQ_A_135458) + (344901*FRQ_U_344901)) / (135458 + 344901)) %>%
  mutate(N = 135458 + 344901) %>%
  rename(Effect_allele = A1, Non_Effect_allele = A2, POS = BP) %>% 
  filter(nchar(Effect_allele) == 1 & Effect_allele %in% c('A', 'C', 'T', 'G')) %>% 
  filter(nchar(Non_Effect_allele) == 1 & Non_Effect_allele %in% c('A', 'C', 'T', 'G')) %>% 
  mutate(Beta = log(OR)) %>%
  mutate(r2 = snp.r2(EAF, Beta)) %>%
  select(SNP, CHR, POS, Effect_allele, Non_Effect_allele, EAF, Beta, SE, P, r2, N)

write_tsv(mdd, gzfile('~/Dropbox/Research/PostDoc-MSSM/2_MR/2_DerivedData/mdd_GWAS.Processed.gz'))

```

```{r}
## Number of Significant SNPs
mdd_ld <- format_data(mdd, type = 'exposure', 
                          phenotype_col = 'trait',
                          snp_col = 'SNP', 
                          beta_col = "Beta",
                          se_col = "SE",
                          eaf_col = "EAF",
                          effect_allele_col = "Effect_allele",
                          other_allele_col = "Non_Effect_allele",
                          pval_col = "P")
mdd_ld <- clump_data(mdd_ld, clump_r2 = 0.1, clump_kb = 250)

clump.n <- mdd_ld %>% 
  mutate(p.cut = cut(pval.exposure, breaks = 
                       c(Inf, 0.05, 5e-5, 5e-6, 5e-7, 5e-8, -Inf), lablels = F)) %>%
  left_join(select(mdd, SNP, r2)) %>% 
  select(SNP, effect_allele.exposure, other_allele.exposure, eaf.exposure, beta.exposure, se.exposure, p.cut, r2) %>% 
  group_by(p.cut) %>% summarise(n = n(), r2 = round(sum(r2, na.rm = T), 4))

all.n <- mdd %>% 
  mutate(p.cut = cut(P, breaks = 
                       c(Inf, 0.05, 5e-5, 5e-6, 5e-7, 5e-8, -Inf), lablels = F)) %>% 
  group_by(p.cut) %>% summarise(n = n(), r2 = round(sum(r2, na.rm = T), 4))

tab.mdd <- full_join(all.n, clump.n, by = 'p.cut', suffix = c('.all', '.clump'))
```

```{r mdd gwas}
p.mdd <- ggman(mdd, snp = 'SNP', chrom = 'CHR', bp = 'POS', pvalue = 'P', ymin = 0, ymax = 20, 
      title = "Major Depressive Disorder - Wray 2018") + 
  theme_classic() + geom_hline(yintercept = -log10(5e-5), colour = 'blue', size = 0.25) + 
  geom_hline(yintercept = -log10(5e-6), colour = 'red', size = 0.25, linetype = 2) 

out.mdd <- grid.arrange(p.mdd,  tableGrob(tab.mdd, theme = ttheme_minimal(base_size = 8)), 
             ncol = 2, widths = 2:1, layout_matrix = rbind(c(1,2), c(1, NA)))

ggsave('~/Dropbox/Research/PostDoc-MSSM/2_MR/4_Output/mdd.png', device = 'png', units = 'in', width = 10, height = 5, plot = out.mdd)
```

## Broad Depression Symptons
[Howard et al 2018](https://www.nature.com/articles/s41467-018-03819-3)): Using UK Biobank participants (113,769 cases and 208,811 controls, ntotal = 322,580, prevalence = 35.27%), broad depression was based on selfreported help-seeking behaviour for mental health difficulties from either a general practitioner or psychiatrist. 

```{r data_wrangling_exposure, echo=F, cache=TRUE}
dep.raw <- read_table2(params$dep.summary)
dep <- dep.raw %>% 
  rename(CHR = chr, SNP = rsid, POS = pos, Effect_allele = A1, Non_Effect_allele = A2, EAF = AlleleFreq, SE = StandardError) %>% 
  filter(nchar(Effect_allele) == 1 & Effect_allele %in% c('A', 'C', 'T', 'G')) %>% 
  filter(nchar(Non_Effect_allele) == 1 & Non_Effect_allele %in% c('A', 'C', 'T', 'G')) %>% 
  mutate(N = 322580) %>% 
  mutate(r2 = snp.r2(EAF, Beta)) %>%
  select(SNP, CHR, POS, Effect_allele, Non_Effect_allele, EAF, Beta, SE, P, r2, N)

write_tsv(dep, gzfile('~/Dropbox/Research/PostDoc-MSSM/2_MR/2_DerivedData/dep_GWAS.Processed.gz'))

```

```{r}
## Number of Significant SNPs
dep_ld <- format_data(filter(dep, P < 5e-5), type = 'exposure', 
                      phenotype_col = 'trait',
                      snp_col = 'SNP', 
                      beta_col = "Beta",
                      se_col = "SE",
                      eaf_col = "EAF",
                      effect_allele_col = "Effect_allele",
                      other_allele_col = "Non_Effect_allele",
                      pval_col = "P")
dep_ld <- clump_data(dep_ld, clump_r2 = 0.1, clump_kb = 250)

clump.n <- dep_ld %>% 
  mutate(p.cut = cut(pval.exposure, breaks = 
                       c(Inf, 0.05, 5e-5, 5e-6, 5e-7, 5e-8, -Inf), lablels = F)) %>%
  left_join(select(dep, SNP, r2)) %>% 
  select(SNP, effect_allele.exposure, other_allele.exposure, eaf.exposure, beta.exposure, se.exposure, p.cut, r2) %>% 
  group_by(p.cut) %>% summarise(n = n(), round(sum(r2, na.rm = T), 4))

all.n <- dep %>% 
  mutate(p.cut = cut(P, breaks = 
                       c(Inf, 0.05, 5e-5, 5e-6, 5e-7, 5e-8, -Inf), lablels = F)) %>% 
  group_by(p.cut) %>% summarise(n = n(), round(sum(r2, na.rm = T), 4))

tab.dep <- full_join(all.n, clump.n, by = 'p.cut', suffix = c('.all', '.clump'))
```

```{r dep gwas}

p.dep <- ggman(filter(dep, P < 0.05), snp = 'SNP', chrom = 'CHR', bp = 'POS', pvalue = 'P', ymin = 0, ymax = 20, 
      title = "Broad Depression Symptons - Howard 2018") + 
  theme_classic() + 
  geom_hline(yintercept = -log10(5e-5), colour = 'blue', size = 0.25) + 
  geom_hline(yintercept = -log10(5e-6), colour = 'red', size = 0.25, linetype = 2) 

out.dep <- grid.arrange(p.dep,  tableGrob(tab.dep, theme = ttheme_minimal(base_size = 8)), 
             ncol = 2, widths = 2:1, layout_matrix = rbind(c(1,2), c(1, NA)))

ggsave('~/Dropbox/Research/PostDoc-MSSM/2_MR/4_Output/dep.png', device = 'png', units = 'in', width = 10, height = 5, plot = out.dep)

```

## Type 2 Diabetes 
[Xu et al Nature Genet 2018](https://www.nature.com/articles/s41467-018-04951-w): Here we conduct a meta-analysis of genome-wide association studies (GWAS) with ~16 million genetic variants in 62,892 T2D cases and 596,424 controls of European ancestry by combining 3 GWAS data sets of European ancestry: DIAbetes Genetics Replication and Meta-analysis (DIAGRAM), Genetic Epidemiology Research on Aging (GERA), and the full cohort release of the UK Biobank (UKB). We identify 139 common and 4 rare variants associated with T2D, 42 of which (39 common and 3 rare variants) are independent of the known variants.

```{r data_wrangling_exposure, echo=F, cache=TRUE}
diab.raw <- read_table2(params$diab.summary)
diab <- diab.raw %>% 
  rename(POS = BP, Effect_allele = A1, Non_Effect_allele = A2, EAF = frq_A1, Beta = b, SE = se) %>% 
  filter(nchar(Effect_allele) == 1 & Effect_allele %in% c('A', 'C', 'T', 'G')) %>% 
  filter(nchar(Non_Effect_allele) == 1 & Non_Effect_allele %in% c('A', 'C', 'T', 'G')) %>%
  mutate(r2 = snp.r2(EAF, Beta)) %>%
  select(SNP, CHR, POS, Effect_allele, Non_Effect_allele, EAF, Beta, SE, P, r2, N)

write_tsv(diab, gzfile('~/Dropbox/Research/PostDoc-MSSM/2_MR/2_DerivedData/diab_GWAS.Processed.gz'))

```

```{r}
## Number of Significant SNPs
diab_ld <- format_data(filter(diab, P < 5e-5), type = 'exposure', 
                      phenotype_col = 'trait',
                      snp_col = 'SNP', 
                      beta_col = "Beta",
                      se_col = "SE",
                      eaf_col = "EAF",
                      effect_allele_col = "Effect_allele",
                      other_allele_col = "Non_Effect_allele",
                      pval_col = "P")
diab_ld <- clump_data(diab_ld, clump_r2 = 0.1, clump_kb = 250)

clump.n <- diab_ld %>% 
  mutate(p.cut = cut(pval.exposure, breaks = 
                       c(Inf, 0.05, 5e-5, 5e-6, 5e-7, 5e-8, -Inf), lablels = F)) %>%
  left_join(select(diab, SNP, r2)) %>% 
  select(SNP, effect_allele.exposure, other_allele.exposure, eaf.exposure, beta.exposure, se.exposure, p.cut, r2) %>% 
  group_by(p.cut) %>% summarise(n = n(), r2 = round(sum(r2, na.rm = T), 4))

all.n <- diab %>% 
  mutate(p.cut = cut(P, breaks = 
                       c(Inf, 0.05, 5e-5, 5e-6, 5e-7, 5e-8, -Inf), lablels = F)) %>% 
  group_by(p.cut) %>% summarise(n = n(), r2 = round(sum(r2, na.rm = T), 4))

tab.diab <- full_join(all.n, clump.n, by = 'p.cut', suffix = c('.all', '.clump'))
```

```{r alc_cons gwas}

p.diab <- ggman(filter(diab, P < 0.05), snp = 'SNP', chrom = 'CHR', bp = 'POS', pvalue = 'P', ymin = 0, ymax = 60, 
      title = "Type 2 Diabetes - Xu 2018") + 
  theme_classic() + 
  geom_hline(yintercept = -log10(5e-5), colour = 'blue', size = 0.25) + 
  geom_hline(yintercept = -log10(5e-6), colour = 'red', size = 0.25, linetype = 2)

out.diab <- grid.arrange(p.diab,  tableGrob(tab.diab, theme = ttheme_minimal(base_size = 8)), 
             ncol = 2, widths = 2:1, layout_matrix = rbind(c(1,2), c(1, NA)))

ggsave('~/Dropbox/Research/PostDoc-MSSM/2_MR/4_Output/diab.png', device = 'png', units = 'in', width = 10, height = 5, plot = out.diab)
```

## Physical Activity
[Klimentidis et al Int J Obes 2018](https://www.nature.com/articles/s41366-018-0120-3): We used data from the UK Biobank to perform the largest genome-wide association study of PA to date, using three measures based on self-report (n = 377,234) and two measures based on wrist-worn accelerometry data (nmax = 91,084).

### Moderate-to-vigorous PA (MVPA) 
Was calculated by taking the sum of total minutes/week of MPA multiplied by four and the total number of VPA minutes/week multiplied by eight, corresponding to their metabolic equivalents, as previously described

```{r data_wrangling_exposure, echo=F, cache=TRUE}
mvpa.raw <- read_table2(params$mvpa.summary)
mvpa <- mvpa.raw %>% 
  rename(POS = BP, Effect_allele = ALLELE1, Non_Effect_allele = ALLELE0, EAF = A1FREQ, Beta = BETA, P = P_BOLT_LMM_INF) %>% 
  filter(nchar(Effect_allele) == 1 & Effect_allele %in% c('A', 'C', 'T', 'G')) %>% 
  filter(nchar(Non_Effect_allele) == 1 & Non_Effect_allele %in% c('A', 'C', 'T', 'G')) %>% 
  mutate(r2 = snp.r2(EAF, Beta)) %>%
  mutate(N = 377234) %>% 
  select(SNP, CHR, POS, Effect_allele, Non_Effect_allele, EAF, Beta, SE, P, r2, N)

write_tsv(mvpa, gzfile('~/Dropbox/Research/PostDoc-MSSM/2_MR/2_DerivedData/mvpa_GWAS.Processed.gz'))

```

```{r}
## Number of Significant SNPs
mvpa_ld <- format_data(filter(mvpa, P < 5e-5), type = 'exposure', 
                       phenotype_col = 'trait',
                       snp_col = 'SNP', 
                       beta_col = "Beta",
                       se_col = "SE",
                       eaf_col = "EAF",
                       effect_allele_col = "Effect_allele",
                       other_allele_col = "Non_Effect_allele",
                       pval_col = "P")
mvpa_ld <- clump_data(mvpa_ld, clump_r2 = 0.1, clump_kb = 250)

clump.n <- mvpa_ld %>% 
  mutate(p.cut = cut(pval.exposure, breaks = 
                       c(Inf, 0.05, 5e-5, 5e-6, 5e-7, 5e-8, -Inf), lablels = F)) %>%
  left_join(select(mvpa, SNP, r2)) %>% 
  select(SNP, effect_allele.exposure, other_allele.exposure, eaf.exposure, beta.exposure, se.exposure, p.cut, r2) %>% 
  group_by(p.cut) %>% summarise(n = n(), r2 = round(sum(r2, na.rm = T), 4))

all.n <- mvpa %>% 
  mutate(p.cut = cut(P, breaks = 
                       c(Inf, 0.05, 5e-5, 5e-6, 5e-7, 5e-8, -Inf), lablels = F)) %>% 
  group_by(p.cut) %>% summarise(n = n(), r2 = round(sum(r2, na.rm = T), 4))

tab.mvpa <- full_join(all.n, clump.n, by = 'p.cut', suffix = c('.all', '.clump'))
```

```{r mvpa gwas}
p.mvpa <- ggman(filter(mvpa, P < 0.05), snp = 'SNP', chrom = 'CHR', bp = 'POS', pvalue = 'P', ymin = 0, ymax = 20, 
                title = "Moderate-to-vigorous PA - Klimentidis 2018") + 
  theme_classic() + 
  geom_hline(yintercept = -log10(5e-5), colour = 'blue', size = 0.25) + 
  geom_hline(yintercept = -log10(5e-6), colour = 'red', size = 0.25, linetype = 2)

out.mvpa <- grid.arrange(p.mvpa,  tableGrob(tab.mvpa, theme = ttheme_minimal(base_size = 8)), 
                         ncol = 2, widths = 2:1, layout_matrix = rbind(c(1,2), c(1, NA)))

ggsave('~/Dropbox/Research/PostDoc-MSSM/2_MR/4_Output/mvpa.png', device = 'png', units = 'in', width = 10, height = 5, plot = out.mvpa)

```

### Vigorous PA
We chose to dichotomize minutes/week of VPA into those who reported 0 days of VPA, and those reporting 3 or more days of VPA and also reporting a typical duration of VPA that is 25 min or greater - 98,060 cases; 162,995 controls
```{r data_wrangling_exposure, echo=F, cache=TRUE}
vpa.raw <- read_table2(params$vpa.summary)
vpa <- vpa.raw %>% 
  rename(POS = BP, Effect_allele = ALLELE1, Non_Effect_allele = ALLELE0, EAF = A1FREQ, Beta = BETA, P = P_BOLT_LMM_INF) %>% 
  filter(nchar(Effect_allele) == 1 & Effect_allele %in% c('A', 'C', 'T', 'G')) %>% 
  filter(nchar(Non_Effect_allele) == 1 & Non_Effect_allele %in% c('A', 'C', 'T', 'G')) %>% 
  mutate(r2 = snp.r2(EAF, Beta)) %>%
  mutate(N = 98060  + 162995) %>% 
  select(SNP, CHR, POS, Effect_allele, Non_Effect_allele, EAF, Beta, SE, P, r2, N)

write_tsv(vpa, gzfile('~/Dropbox/Research/PostDoc-MSSM/2_MR/2_DerivedData/vpa_GWAS.Processed.gz'))

```

```{r}
## Number of Significant SNPs
vpa_ld <- format_data(filter(vpa, P < 5e-5), type = 'exposure', 
                         phenotype_col = 'trait',
                         snp_col = 'SNP', 
                         beta_col = "Beta",
                         se_col = "SE",
                         eaf_col = "EAF",
                         effect_allele_col = "Effect_allele",
                         other_allele_col = "Non_Effect_allele",
                         pval_col = "P")
vpa_ld <- clump_data(vpa_ld, clump_r2 = 0.1, clump_kb = 250)

clump.n <- vpa_ld %>% 
  mutate(p.cut = cut(pval.exposure, breaks = 
                       c(Inf, 0.05, 5e-5, 5e-6, 5e-7, 5e-8, -Inf), lablels = F)) %>%
  left_join(select(vpa, SNP, r2)) %>% 
  select(SNP, effect_allele.exposure, other_allele.exposure, eaf.exposure, beta.exposure, se.exposure, p.cut, r2) %>% 
  group_by(p.cut) %>% summarise(n = n(), r2 = round(sum(r2, na.rm = T), 4))

all.n <- vpa %>% 
  mutate(p.cut = cut(P, breaks = 
                       c(Inf, 0.05, 5e-5, 5e-6, 5e-7, 5e-8, -Inf), lablels = F)) %>% 
  group_by(p.cut) %>% summarise(n = n(), r2 = round(sum(r2, na.rm = T), 4))

tab.vpa <- full_join(all.n, clump.n, by = 'p.cut', suffix = c('.all', '.clump'))
```

```{r alc_cons gwas}

p.vpa <- ggman(filter(vpa, P < 0.05), snp = 'SNP', chrom = 'CHR', bp = 'POS', pvalue = 'P', ymin = 0, ymax = 20, 
                title = "Vigorous PA - Klimentidis 2018") + 
  theme_classic() + 
  geom_hline(yintercept = -log10(5e-5), colour = 'blue', size = 0.25) + 
  geom_hline(yintercept = -log10(5e-6), colour = 'red', size = 0.25, linetype = 2)

out.vpa <- grid.arrange(p.vpa,  tableGrob(tab.vpa, theme = ttheme_minimal(base_size = 8)), 
                         ncol = 2, widths = 2:1, layout_matrix = rbind(c(1,2), c(1, NA)))

ggsave('~/Dropbox/Research/PostDoc-MSSM/2_MR/4_Output/vpa.png', device = 'png', units = 'in', width = 10, height = 5, plot = out.vpa)

```

### Strenuous sports or other exercises
We identified individuals spending 2–3 days/week or more doing strenuous sports or other exercises (SSOE), for a duration of 15–30 min or greater. Controls were those individuals who did not indicate spending any time in the last 4 weeks doing either SSOE - 124,842 cases; 225,650 controls
```{r data_wrangling_exposure, echo=F, cache=TRUE}
ssoe.raw <- read_table2(params$ssoe.summary)
ssoe <- ssoe.raw %>% 
  rename(POS = BP, Effect_allele = ALLELE1, Non_Effect_allele = ALLELE0, EAF = A1FREQ, Beta = BETA, P = P_BOLT_LMM_INF) %>% 
  filter(nchar(Effect_allele) == 1 & Effect_allele %in% c('A', 'C', 'T', 'G')) %>% 
  filter(nchar(Non_Effect_allele) == 1 & Non_Effect_allele %in% c('A', 'C', 'T', 'G')) %>% 
  mutate(r2 = snp.r2(EAF, Beta)) %>%
  mutate(N = 124842 + 225650) %>% 
  select(SNP, CHR, POS, Effect_allele, Non_Effect_allele, EAF, Beta, SE, P, r2, N)

write_tsv(ssoe, gzfile('~/Dropbox/Research/PostDoc-MSSM/2_MR/2_DerivedData/ssoe_GWAS.Processed.gz'))

```

```{r}
## Number of Significant SNPs
ssoe_ld <- format_data(filter(ssoe, P < 5e-5), type = 'exposure', 
                         phenotype_col = 'trait',
                         snp_col = 'SNP', 
                         beta_col = "Beta",
                         se_col = "SE",
                         eaf_col = "EAF",
                         effect_allele_col = "Effect_allele",
                         other_allele_col = "Non_Effect_allele",
                         pval_col = "P")
ssoe_ld <- clump_data(ssoe_ld, clump_r2 = 0.1, clump_kb = 250)

clump.n <- ssoe_ld %>% 
  mutate(p.cut = cut(pval.exposure, breaks = 
                       c(Inf, 0.05, 5e-5, 5e-6, 5e-7, 5e-8, -Inf), lablels = F)) %>%
  left_join(select(ssoe, SNP, r2)) %>% 
  select(SNP, effect_allele.exposure, other_allele.exposure, eaf.exposure, beta.exposure, se.exposure, p.cut, r2) %>% 
  group_by(p.cut) %>% summarise(n = n(), r2 = round(sum(r2, na.rm = T), 4))

all.n <- ssoe %>% 
  mutate(p.cut = cut(P, breaks = 
                       c(Inf, 0.05, 5e-5, 5e-6, 5e-7, 5e-8, -Inf), lablels = F)) %>% 
  group_by(p.cut) %>% summarise(n = n(), r2 = round(sum(r2, na.rm = T), 4))

tab.ssoe <- full_join(all.n, clump.n, by = 'p.cut', suffix = c('.all', '.clump'))
```

```{r ssoe gwas}

p.ssoe <- ggman(filter(ssoe, P < 0.05), snp = 'SNP', chrom = 'CHR', bp = 'POS', pvalue = 'P', ymin = 0, ymax = 20, 
                title = "SSOE - Klimentidis 2018") + 
  theme_classic() + 
  geom_hline(yintercept = -log10(5e-5), colour = 'blue', size = 0.25) + 
  geom_hline(yintercept = -log10(5e-6), colour = 'red', size = 0.25, linetype = 2)

out.ssoe <- grid.arrange(p.ssoe,  tableGrob(tab.ssoe, theme = ttheme_minimal(base_size = 8)), 
                         ncol = 2, widths = 2:1, layout_matrix = rbind(c(1,2), c(1, NA)))

ggsave('~/Dropbox/Research/PostDoc-MSSM/2_MR/4_Output/ssoe.png', device = 'png', units = 'in', width = 10, height = 5, plot = out.ssoe)

```

### Accelerometry – Fraction accelerations > 425 milli-gravities
Also, in the UK Biobank, approximately 100,000 participants wore an Axivity AX3 wrist-worn accelerometer, as previously described. We examined two measures derived from up to seven days of accelerometer wear: overall acceleration average, and fraction of accelerations >425 milli-gravities (mg). The 425 mg cutoff was chosen because this corresponds to an equivalent of vigorous PA (6 METs).  
```{r data_wrangling_exposure, echo=F, cache=TRUE}
acc425.raw <- read_table2(params$acc425.summary)
acc425 <- acc425.raw %>% 
  rename(POS = BP, Effect_allele = ALLELE1, Non_Effect_allele = ALLELE0, EAF = A1FREQ, Beta = BETA, P = P_BOLT_LMM_INF) %>% 
  filter(nchar(Effect_allele) == 1 & Effect_allele %in% c('A', 'C', 'T', 'G')) %>% 
  filter(nchar(Non_Effect_allele) == 1 & Non_Effect_allele %in% c('A', 'C', 'T', 'G')) %>% 
  mutate(r2 = snp.r2(EAF, Beta)) %>%
  mutate(N = 91084) %>% 
  select(SNP, CHR, POS, Effect_allele, Non_Effect_allele, EAF, Beta, SE, P, r2, N)

write_tsv(acc425, gzfile('~/Dropbox/Research/PostDoc-MSSM/2_MR/2_DerivedData/acc425_GWAS.Processed.gz'))

```

```{r}
## Number of Significant SNPs
acc425_ld <- format_data(filter(acc425, P < 5e-5), type = 'exposure', 
                       phenotype_col = 'trait',
                       snp_col = 'SNP', 
                       beta_col = "Beta",
                       se_col = "SE",
                       eaf_col = "EAF",
                       effect_allele_col = "Effect_allele",
                       other_allele_col = "Non_Effect_allele",
                       pval_col = "P")
acc425_ld <- clump_data(acc425_ld, clump_r2 = 0.1, clump_kb = 250)

clump.n <- acc425_ld %>% 
  mutate(p.cut = cut(pval.exposure, breaks = 
                       c(Inf, 0.05, 5e-5, 5e-6, 5e-7, 5e-8, -Inf), lablels = F)) %>%
  left_join(select(acc425, SNP, r2)) %>% 
  select(SNP, effect_allele.exposure, other_allele.exposure, eaf.exposure, beta.exposure, se.exposure, p.cut, r2) %>% 
  group_by(p.cut) %>% summarise(n = n(), r2 = round(sum(r2, na.rm = T), 4))

all.n <- acc425 %>% 
  mutate(p.cut = cut(P, breaks = 
                       c(Inf, 0.05, 5e-5, 5e-6, 5e-7, 5e-8, -Inf), lablels = F)) %>% 
  group_by(p.cut) %>% summarise(n = n(), r2 = round(sum(r2, na.rm = T), 4))

tab.acc425 <- full_join(all.n, clump.n, by = 'p.cut', suffix = c('.all', '.clump'))
full_join(all.n, clump.n, by = 'p.cut', suffix = c('.all', '.clump'))
```

```{r acc425 gwas}
p.acc425 <- ggman(filter(acc425, P < 0.05), snp = 'SNP', chrom = 'CHR', bp = 'POS', pvalue = 'P', ymin = 0, ymax = 15, 
                title = "Accelerations > 425 milli-gravities - Klimentidis 2018") + 
  theme_classic() + 
  geom_hline(yintercept = -log10(5e-5), colour = 'blue', size = 0.25) + 
  geom_hline(yintercept = -log10(5e-6), colour = 'red', size = 0.25, linetype = 2)

out.acc425 <- grid.arrange(p.acc425,  tableGrob(tab.acc425, theme = ttheme_minimal(base_size = 8)), 
                         ncol = 2, widths = 2:1, layout_matrix = rbind(c(1,2), c(1, NA)))

ggsave('~/Dropbox/Research/PostDoc-MSSM/2_MR/4_Output/acc425.png', device = 'png', units = 'in', width = 10, height = 5, plot = out.acc425)

```

### Accelerometry – Average acceleration (milli-gravities)
Also, in the UK Biobank, approximately 100,000 participants wore an Axivity AX3 wrist-worn accelerometer, as previously described. We examined two measures derived from up to seven days of accelerometer wear: overall acceleration average, and fraction of accelerations >425 milli-gravities (mg). 
```{r data_wrangling_exposure, echo=F, cache=TRUE}
accave.raw <- read_table2(params$accave.summary)
accave <- accave.raw %>% 
  rename(POS = BP, Effect_allele = ALLELE1, Non_Effect_allele = ALLELE0, EAF = A1FREQ, Beta = BETA, P = P_BOLT_LMM_INF) %>% 
  filter(nchar(Effect_allele) == 1 & Effect_allele %in% c('A', 'C', 'T', 'G')) %>% 
  filter(nchar(Non_Effect_allele) == 1 & Non_Effect_allele %in% c('A', 'C', 'T', 'G')) %>% 
  mutate(r2 = snp.r2(EAF, Beta)) %>%
  mutate(N = 90667) %>% 
  select(SNP, CHR, POS, Effect_allele, Non_Effect_allele, EAF, Beta, SE, P, r2, N)

write_tsv(accave, gzfile('~/Dropbox/Research/PostDoc-MSSM/2_MR/2_DerivedData/accave_GWAS.Processed.gz'))

```

```{r}
## Number of Significant SNPs
accave_ld <- format_data(filter(accave, P < 5e-5), type = 'exposure', 
                       phenotype_col = 'trait',
                       snp_col = 'SNP', 
                       beta_col = "Beta",
                       se_col = "SE",
                       eaf_col = "EAF",
                       effect_allele_col = "Effect_allele",
                       other_allele_col = "Non_Effect_allele",
                       pval_col = "P")
accave_ld <- clump_data(accave_ld, clump_r2 = 0.1, clump_kb = 250)

clump.n <- accave_ld %>% 
  mutate(p.cut = cut(pval.exposure, breaks = 
                       c(Inf, 0.05, 5e-5, 5e-6, 5e-7, 5e-8, -Inf), lablels = F)) %>%
  left_join(select(accave, SNP, r2)) %>% 
  select(SNP, effect_allele.exposure, other_allele.exposure, eaf.exposure, beta.exposure, se.exposure, p.cut, r2) %>% 
  group_by(p.cut) %>% summarise(n = n(), r2 = round(sum(r2, na.rm = T), 4))

all.n <- accave %>% 
  mutate(p.cut = cut(P, breaks = 
                       c(Inf, 0.05, 5e-5, 5e-6, 5e-7, 5e-8, -Inf), lablels = F)) %>% 
  group_by(p.cut) %>% summarise(n = n(), r2 = round(sum(r2, na.rm = T), 4))

tab.accave <- full_join(all.n, clump.n, by = 'p.cut', suffix = c('.all', '.clump'))
```

```{r alc_cons gwas}

p.accave <- ggman(filter(accave, P < 0.05), snp = 'SNP', chrom = 'CHR', bp = 'POS', pvalue = 'P', ymin = 0, ymax = 20, 
                title = "Average acceleration - Klimentidis 2018") + 
  theme_classic() + 
  geom_hline(yintercept = -log10(5e-5), colour = 'blue', size = 0.25) + 
  geom_hline(yintercept = -log10(5e-6), colour = 'red', size = 0.25, linetype = 2)

out.accave <- grid.arrange(p.accave,  tableGrob(tab.accave, theme = ttheme_minimal(base_size = 8)), 
                         ncol = 2, widths = 2:1, layout_matrix = rbind(c(1,2), c(1, NA)))

ggsave('~/Dropbox/Research/PostDoc-MSSM/2_MR/4_Output/accave.png', device = 'png', units = 'in', width = 10, height = 5, plot = out.accave)

```

### Educational Attainment 
[Lee et al Nature Genet 2018](https://www.nature.com/articles/s41588-018-0147-3) Here we conducted a large-scale genetic association analysis of educational attainment in a sample of approximately 1.1 million individuals and identify 1,271 independent genome-wide-significant SNPs. 

```{r data_wrangling_exposure, echo=F, cache=TRUE}
educ.raw <- read_table2(params$edu10k.summary)
educ <- educ.raw %>% 
  rename(SNP = MarkerName, Effect_allele = A1, Non_Effect_allele = A2, P = Pval) %>% 
  filter(nchar(Effect_allele) == 1 & Effect_allele %in% c('A', 'C', 'T', 'G')) %>% 
  filter(nchar(Non_Effect_allele) == 1 & Non_Effect_allele %in% c('A', 'C', 'T', 'G')) %>% 
  mutate(r2 = snp.r2(EAF, Beta)) %>%
  mutate(N = 1131881) %>% 
  select(SNP, CHR, POS, Effect_allele, Non_Effect_allele, EAF, Beta, SE, P, r2, N)

write_tsv(educ, gzfile('~/Dropbox/Research/PostDoc-MSSM/2_MR/2_DerivedData/educ_GWAS.Processed.gz'))

```

```{r}
## Number of Significant SNPs
educ_ld <- format_data(educ, type = 'exposure', 
                       phenotype_col = 'trait',
                       snp_col = 'SNP', 
                       beta_col = "Beta",
                       se_col = "SE",
                       eaf_col = "EAF",
                       effect_allele_col = "Effect_allele",
                       other_allele_col = "Non_Effect_allele",
                       pval_col = "P")
educ_ld <- clump_data(educ_ld, clump_r2 = 0.1, clump_kb = 250)

clump.n <- educ_ld %>% 
  mutate(p.cut = cut(pval.exposure, breaks = 
                       c(Inf, 0.05, 5e-5, 5e-6, 5e-7, 5e-8, -Inf), lablels = F)) %>%
  left_join(select(educ, SNP, r2)) %>% 
  select(SNP, effect_allele.exposure, other_allele.exposure, eaf.exposure, beta.exposure, se.exposure, p.cut, r2) %>% 
  group_by(p.cut) %>% summarise(n = n(), r2 = round(sum(r2, na.rm = T), 4))

all.n <- educ %>% 
  mutate(p.cut = cut(P, breaks = 
                       c(Inf, 0.05, 5e-5, 5e-6, 5e-7, 5e-8, -Inf), lablels = F)) %>% 
  group_by(p.cut) %>% summarise(n = n(), r2 = round(sum(r2, na.rm = T), 4))

tab.educ <- full_join(all.n, clump.n, by = 'p.cut', suffix = c('.all', '.clump'))
```

```{r alc_cons gwas}

p.educ <- ggman(filter(educ, P < 0.05), snp = 'SNP', chrom = 'CHR', bp = 'POS', pvalue = 'P', ymin = 0, ymax = 60, 
                title = "Educational Attainment - Lee 2018") + 
  theme_classic() + 
  geom_hline(yintercept = -log10(5e-5), colour = 'blue', size = 0.25) + 
  geom_hline(yintercept = -log10(5e-6), colour = 'red', size = 0.25, linetype = 2)

out.educ <- arrangeGrob(p.educ,  tableGrob(tab.educ, theme = ttheme_minimal(base_size = 8)), 
                         ncol = 2, widths = 2:1, layout_matrix = rbind(c(1,2), c(1, NA)))

ggsave('~/Dropbox/Research/PostDoc-MSSM/2_MR/4_Output/educ.png', device = 'png', units = 'in', width = 10, height = 5, plot = out.educ)

```

## Smoking UKB 
```{r SMK_UKB}
## Gene Atlas
path <- '~/Dropbox/Research/Data/Summary_Statisitics/Smoking/ukbb/'
filenames <- list.files(path, pattern="*.csv.gz")

## effects
files <- paste(path, filenames[grep('allWhites', filenames)], sep="")
out <- lapply(files, read_table2)
out <- do.call(rbind, out)

## snp info
files <- paste(path, filenames[grep('snps.imputed', filenames)], sep="")
info <- lapply(files, read_table2)
info <- do.call(rbind, info)

smkukbb <- out %>% 
  left_join(info, by = 'SNP') %>% 
  left_join(hrc, by = c('SNP' = 'ID')) %>% 
  filter(!is.na(CHROM)) %>% 
  filter(`HWE-P` > 1e-6) %>% 
  filter(MAF > 1e-4) %>% 
  filter(iscores >= 0.8) %>%
  mutate(A2 = ifelse(ALLELE == ALT, REF, ALLELE)) %>% 
  rename(CHR = CHROM, Effect_allele = ALLELE, Non_Effect_allele = A2, Beta = `NBETA-20116-0.0`, SE = `NSE-20116-0.0`, P = `PV-20116-0.0`, EAF = MAF) %>% 
  mutate(r2 = snp.r2(EAF, Beta)) %>%
  select(CHR, POS, SNP, Effect_allele, Non_Effect_allele, EAF, Beta, SE, P, r2)

write_tsv(smkukbb, gzfile('~/Dropbox/Research/PostDoc-MSSM/2_MR/2_DerivedData/smkukbb_GWAS.Processed.gz'))  
```

```{r}
## Number of Significant SNPs
smkukbb_ld <- format_data(filter(smkukbb, P < 5e-5), type = 'exposure', 
                               phenotype_col = 'trait',
                               snp_col = 'SNP', 
                               beta_col = "Beta",
                               se_col = "SE",
                               eaf_col = "EAF",
                               effect_allele_col = "Effect_allele",
                               other_allele_col = "Non_Effect_allele",
                               pval_col = "P")
smkukbb_ld <- clump_data(smkukbb_ld, clump_r2 = 0.1, clump_kb = 250)

clump.n <- smkukbb_ld %>% 
  mutate(p.cut = cut(pval.exposure, breaks = 
                       c(Inf, 0.05, 5e-5, 5e-6, 5e-7, 5e-8, -Inf), lablels = F)) %>%
  left_join(select(smkukbb, SNP, r2)) %>% 
  select(SNP, effect_allele.exposure, other_allele.exposure, eaf.exposure, beta.exposure, se.exposure, p.cut, r2) %>% 
  group_by(p.cut) %>% summarise(n = n(), r2 = round(sum(r2, na.rm = T), 4))

all.n <- smkukbb %>% 
  mutate(p.cut = cut(P, breaks = 
                       c(Inf, 0.05, 5e-5, 5e-6, 5e-7, 5e-8, -Inf), lablels = F)) %>% 
  group_by(p.cut) %>% summarise(n = n(), r2 = round(sum(r2, na.rm = T), 4))

tab.smkukbb <- full_join(all.n, clump.n, by = 'p.cut', suffix = c('.all', '.clump'))
```

```{r smkukbb gwas}
## Plots
p.smkukbb <- ggman(filter(smkukbb, P < 0.05), snp = 'SNP', chrom = 'CHR', bp = 'POS', pvalue = 'P', ymin = 0, ymax = 30, 
                title = "Smoking UKBB") + 
  theme_classic() + 
  geom_hline(yintercept = -log10(5e-5), colour = 'blue', size = 0.25) + 
  geom_hline(yintercept = -log10(5e-6), colour = 'red', size = 0.25, linetype = 2)

out.smkukbb <- grid.arrange(p.smkukbb,  tableGrob(tab.smkukbb, theme = ttheme_minimal(base_size = 8)), 
                         ncol = 2, widths = 2:1, layout_matrix = rbind(c(1,2), c(1, NA)))

ggsave('~/Dropbox/Research/PostDoc-MSSM/2_MR/4_Output/smkukbb.png', device = 'png', units = 'in', width = 10, height = 5, plot = out.smkukbb)

```

## Smoking TAG
[TAG Nature Genet 2010](https://www.nature.com/articles/ng.571): TAG conducted a meta-analysis of Ever vs never smokers in 74,053 participants.

```{r data_wrangling_exposure, echo=F, cache=TRUE}
evrsmk.raw <- read_table2(params$evrsmk.summary) 

evrsmk_hg18 <- evrsmk.raw %>% 
  mutate(chrom = paste0('chr', CHR)) %>% 
  mutate(chromEnd = as.integer(BP + 1)) %>% 
  rename(chromStart = BP, name = SNP) %>% 
  select(chrom, chromStart, chromEnd, name) %>% 
  write_delim('~/Dropbox/Research/Data/Summary_Statisitics/Smoking/tag/evrsmk_hg18.BED', delim = " ", col_names = F)

system('/Users/sheaandrews/Programs/liftOver/liftOver /Users/sheaandrews/Dropbox/Research/Data/Summary_Statisitics/Smoking/tag/evrsmk_hg18.BED /Users/sheaandrews/Programs/liftOver/hg18ToHg19.over.chain.gz  /Users/sheaandrews/Dropbox/Research/Data/Summary_Statisitics/Smoking/tag/evrsmk_hg19.BED  /Users/sheaandrews/Dropbox/Research/Data/Summary_Statisitics/Smoking/tag/evrsmk_hg19_unlifted.bed')

evrsmk <- read_table2('/Users/sheaandrews/Dropbox/Research/Data/Summary_Statisitics/Smoking/tag/evrsmk_hg19.BED', col_names = F) %>% 
  rename(POS_hg19 = X2, SNP = X4) %>% 
  select(POS_hg19, SNP) %>% 
  left_join(evrsmk.raw, by = 'SNP') %>% 
  rename(Effect_allele = A1, Non_Effect_allele = A2, Beta = OR, POS = POS_hg19, EAF = FRQ_A) %>% 
  mutate(r2 = snp.r2(EAF, Beta)) %>%
  mutate(N = 74053) %>% 
  select(SNP, CHR, POS, Effect_allele, Non_Effect_allele, EAF, Beta, SE, P, r2, N)

write_tsv(evrsmk, gzfile('~/Dropbox/Research/PostDoc-MSSM/2_MR/2_DerivedData/evrsmk_GWAS.Processed.gz'))

```

```{r}
## Number of Significant SNPs
evrsmk_ld <- format_data(filter(evrsmk, P < 5e-5), type = 'exposure', 
                               phenotype_col = 'trait',
                               snp_col = 'SNP', 
                               beta_col = "Beta",
                               se_col = "SE",
                               eaf_col = "EAF",
                               effect_allele_col = "Effect_allele",
                               other_allele_col = "Non_Effect_allele",
                               pval_col = "P")
evrsmk_ld <- clump_data(evrsmk_ld, clump_r2 = 0.1, clump_kb = 250)

clump.n <- evrsmk_ld %>% 
  mutate(p.cut = cut(pval.exposure, breaks = 
                       c(Inf, 0.05, 5e-5, 5e-6, 5e-7, 5e-8, -Inf), lablels = F)) %>%
  left_join(select(evrsmk, SNP, r2)) %>% 
  select(SNP, effect_allele.exposure, other_allele.exposure, eaf.exposure, beta.exposure, se.exposure, p.cut, r2) %>% 
  group_by(p.cut) %>% summarise(n = n(), r2 = round(sum(r2, na.rm = T), 4))

all.n <- evrsmk %>% 
  mutate(p.cut = cut(P, breaks = 
                       c(Inf, 0.05, 5e-5, 5e-6, 5e-7, 5e-8, -Inf), lablels = F)) %>% 
  group_by(p.cut) %>% summarise(n = n(), r2 = round(sum(r2, na.rm = T), 4))

tab.evrsmk <- full_join(all.n, clump.n, by = 'p.cut', suffix = c('.all', '.clump'))
```

```{r evrsmk gwas}
## Plots
p.evrsmk <- ggman(filter(evrsmk, P < 0.05), snp = 'SNP', chrom = 'CHR', bp = 'POS', pvalue = 'P', ymin = 0, ymax = 10, 
                title = "Ever Smoker - TAG 2010") + 
  theme_classic() + 
  geom_hline(yintercept = -log10(5e-5), colour = 'blue', size = 0.25) + 
  geom_hline(yintercept = -log10(5e-6), colour = 'red', size = 0.25, linetype = 2)

out.evrsmk <- grid.arrange(p.evrsmk,  tableGrob(tab.evrsmk, theme = ttheme_minimal(base_size = 8)), 
                         ncol = 2, widths = 2:1, layout_matrix = rbind(c(1,2), c(1, NA)))

ggsave('~/Dropbox/Research/PostDoc-MSSM/2_MR/4_Output/evrsmk.png', device = 'png', units = 'in', width = 10, height = 5, plot = out.evrsmk)

```



## Oily Fish UKB 
```{r fish_UKB}
path <- '~/Dropbox/Research/Data/Summary_Statisitics/FishIntake/'
filenames <- list.files(path, pattern="*.csv.gz")

## effects
files <- paste(path, filenames[grep('allWhites', filenames)], sep="")
out <- lapply(files, read_table2)
out <- do.call(rbind, out)

## snp info
files <- paste(path, filenames[grep('snps.imputed', filenames)], sep="")
info <- lapply(files, read_table2)
info <- do.call(rbind, info)

fish <- out %>% 
  left_join(info, by = 'SNP') %>% 
  left_join(hrc, by = c('SNP' = 'ID')) %>% 
  filter(!is.na(CHROM)) %>% 
  filter(`HWE-P` > 1e-6) %>% 
  filter(MAF > 1e-4) %>% 
  filter(iscores >= 0.8) %>%
  mutate(A2 = ifelse(ALLELE == ALT, REF, ALLELE)) %>% 
  rename(CHR = CHROM, Effect_allele = ALLELE, Non_Effect_allele = A2, Beta = `NBETA-1329-0.0`, SE = `NSE-1329-0.0`, P = `PV-1329-0.0`, EAF = MAF) %>% 
  mutate(r2 = snp.r2(EAF, Beta)) %>%
  select(CHR, POS, SNP, Effect_allele, Non_Effect_allele, EAF, Beta, SE, P, r2)

write_tsv(fish, gzfile('~/Dropbox/Research/PostDoc-MSSM/2_MR/2_DerivedData/fish_GWAS.Processed.gz'))    
```

```{r}
## Number of Significant SNPs
fish_ld <- format_data(filter(fish, P < 5e-5), type = 'exposure', 
                               phenotype_col = 'trait',
                               snp_col = 'SNP', 
                               beta_col = "Beta",
                               se_col = "SE",
                               eaf_col = "EAF",
                               effect_allele_col = "Effect_allele",
                               other_allele_col = "Non_Effect_allele",
                               pval_col = "P")
fish_ld <- clump_data(fish_ld, clump_r2 = 0.1, clump_kb = 250)

clump.n <- fish_ld %>% 
  mutate(p.cut = cut(pval.exposure, breaks = 
                       c(Inf, 0.05, 5e-5, 5e-6, 5e-7, 5e-8, -Inf), lablels = F)) %>%
  left_join(select(fish, SNP, r2)) %>% 
  select(SNP, effect_allele.exposure, other_allele.exposure, eaf.exposure, beta.exposure, se.exposure, p.cut, r2) %>% 
  group_by(p.cut) %>% summarise(n = n(), r2 = round(sum(r2, na.rm = T), 4))

all.n <- fish %>% 
  mutate(p.cut = cut(P, breaks = 
                       c(Inf, 0.05, 5e-5, 5e-6, 5e-7, 5e-8, -Inf), lablels = F)) %>% 
  group_by(p.cut) %>% summarise(n = n(), r2 = round(sum(r2, na.rm = T), 4))

tab.fish <- full_join(all.n, clump.n, by = 'p.cut', suffix = c('.all', '.clump'))
```

```{r fish gwas}
## Plots
p.fish <- ggman(filter(fish, P < 0.05), snp = 'SNP', chrom = 'CHR', bp = 'POS', pvalue = 'P', ymin = 0, ymax = 25, 
                title = "Oily Fish Intake") + 
  theme_classic() + 
  geom_hline(yintercept = -log10(5e-5), colour = 'blue', size = 0.25) + 
  geom_hline(yintercept = -log10(5e-6), colour = 'red', size = 0.25, linetype = 2)

out.fish <- grid.arrange(p.fish,  tableGrob(tab.fish, theme = ttheme_minimal(base_size = 8)), 
                         ncol = 2, widths = 2:1, layout_matrix = rbind(c(1,2), c(1, NA)))

ggsave('~/Dropbox/Research/PostDoc-MSSM/2_MR/4_Output/fish.png', device = 'png', units = 'in', width = 10, height = 5, plot = out.fish)

```

## HDL Cholesterol 
[Willer et al Nat Genet 2013](https://www.nature.com/articles/ng.2797): To identify new loci and refine known loci influencing these lipids, we examined 188,577 individuals using genome-wide and custom genotyping arrays. We identify and annotate 157 loci associated with lipid levels at P < 5 × 10−8, including 62 loci not previously associated with lipid levels in humans. 
```{r data_wrangling_exposure, echo=F, cache=TRUE}
hdl.raw <- read_table2(params$hdl.summary)
hdl <- hdl.raw %>% 
  rename(SNP = rsid, Effect_allele = A1, Non_Effect_allele = A2, P = `P-value`, EAF = Freq.A1.1000G.EUR, Beta = beta, SE = se) %>% 
  separate(SNP_hg19, c('CHR', 'POS'), sep = ':') %>%
  mutate(CHR = str_replace(CHR, 'chr', '')) %>% 
  mutate(Effect_allele = toupper(Effect_allele), Non_Effect_allele = toupper(Non_Effect_allele)) %>%
  filter(nchar(Effect_allele) == 1 & Effect_allele %in% c('A', 'C', 'T', 'G')) %>% 
  filter(nchar(Non_Effect_allele) == 1 & Non_Effect_allele %in% c('A', 'C', 'T', 'G')) %>% 
  mutate(r2 = snp.r2(EAF, Beta)) %>%
  select(SNP, CHR, POS, Effect_allele, Non_Effect_allele, EAF, Beta, SE, P, r2, N)

write_tsv(hdl, gzfile('~/Dropbox/Research/PostDoc-MSSM/2_MR/2_DerivedData/hdl_GWAS.Processed.gz'))

```

```{r}
## Number of Significant SNPs
hdl_ld <- format_data(filter(hdl, P < 5e-5), type = 'exposure', 
                       phenotype_col = 'hdl',
                       snp_col = 'SNP', 
                       beta_col = "Beta",
                       se_col = "SE",
                       eaf_col = "EAF",
                       effect_allele_col = "Effect_allele",
                       other_allele_col = "Non_Effect_allele",
                       pval_col = "P")
hdl_ld <- clump_data(hdl_ld, clump_r2 = 0.1, clump_kb = 250)

clump.n <- hdl_ld %>% 
  mutate(p.cut = cut(pval.exposure, breaks = 
                       c(Inf, 0.05, 5e-5, 5e-6, 5e-7, 5e-8, -Inf), lablels = F)) %>%
  left_join(select(hdl, SNP, r2)) %>% 
  select(SNP, effect_allele.exposure, other_allele.exposure, eaf.exposure, beta.exposure, se.exposure, p.cut, r2) %>% 
  group_by(p.cut) %>% summarise(n = n(), r2 = round(sum(r2, na.rm = T), 4))

all.n <- hdl %>% 
  mutate(p.cut = cut(P, breaks = 
                       c(Inf, 0.05, 5e-5, 5e-6, 5e-7, 5e-8, -Inf), lablels = F)) %>% 
  group_by(p.cut) %>% summarise(n = n(), r2 = round(sum(r2, na.rm = T), 4))

tab.hdl <- full_join(all.n, clump.n, by = 'p.cut', suffix = c('.all', '.clump'))
```

```{r alc_cons gwas}

p.hdl <- ggman(filter(hdl, P < 0.05), snp = 'SNP', chrom = 'CHR', bp = 'POS', pvalue = 'P', ymin = 0, ymax = 60, 
                title = "HDL Cholesterol - Willer 2013") + 
  theme_classic() + 
  geom_hline(yintercept = -log10(5e-5), colour = 'blue', size = 0.25) + 
  geom_hline(yintercept = -log10(5e-6), colour = 'red', size = 0.25, linetype = 2)

out.hdl <- grid.arrange(p.hdl,  tableGrob(tab.hdl, theme = ttheme_minimal(base_size = 8)), 
                         ncol = 2, widths = 2:1, layout_matrix = rbind(c(1,2), c(1, NA)))

ggsave('~/Dropbox/Research/PostDoc-MSSM/2_MR/4_Output/hdl.png', device = 'png', units = 'in', width = 10, height = 5, plot = out.hdl)

```


## LDL Cholesterol 
[Willer et al Nat Genet 2013](https://www.nature.com/articles/ng.2797): To identify new loci and refine known loci influencing these lipids, we examined 188,577 individuals using genome-wide and custom genotyping arrays. We identify and annotate 157 loci associated with lipid levels at P < 5 × 10−8, including 62 loci not previously associated with lipid levels in humans. 
```{r data_wrangling_exposure, echo=F, cache=TRUE}
ldl.raw <- read_table2(params$ldl.summary)
ldl <- ldl.raw %>% 
  rename(SNP = rsid, Effect_allele = A1, Non_Effect_allele = A2, P = `P-value`, EAF = Freq.A1.1000G.EUR, Beta = beta, SE = se) %>% 
  separate(SNP_hg19, c('CHR', 'POS'), sep = ':') %>%
  mutate(CHR = as.numeric(str_replace(CHR, 'chr', ''))) %>% 
  arrange(CHR, POS) %>%
  mutate(Effect_allele = toupper(Effect_allele), Non_Effect_allele = toupper(Non_Effect_allele)) %>%
  filter(nchar(Effect_allele) == 1 & Effect_allele %in% c('A', 'C', 'T', 'G')) %>% 
  filter(nchar(Non_Effect_allele) == 1 & Non_Effect_allele %in% c('A', 'C', 'T', 'G')) %>% 
  mutate(r2 = snp.r2(EAF, Beta)) %>%
  select(SNP, CHR, POS, Effect_allele, Non_Effect_allele, EAF, Beta, SE, P, r2, N)

write_tsv(ldl, gzfile('~/Dropbox/Research/PostDoc-MSSM/2_MR/2_DerivedData/ldl_GWAS.Processed.gz'))

```

```{r}
## Number of Significant SNPs
ldl_ld <- format_data(filter(ldl, P < 5e-5), type = 'exposure', 
                      phenotype_col = 'ldl',
                      snp_col = 'SNP', 
                      beta_col = "Beta",
                      se_col = "SE",
                      eaf_col = "EAF",
                      effect_allele_col = "Effect_allele",
                      other_allele_col = "Non_Effect_allele",
                      pval_col = "P")
ldl_ld <- clump_data(ldl_ld, clump_r2 = 0.1, clump_kb = 250)

clump.n <- ldl_ld %>% 
  mutate(p.cut = cut(pval.exposure, breaks = 
                       c(Inf, 0.05, 5e-5, 5e-6, 5e-7, 5e-8, -Inf), lablels = F)) %>%
  left_join(select(ldl, SNP, r2)) %>% 
  select(SNP, effect_allele.exposure, other_allele.exposure, eaf.exposure, beta.exposure, se.exposure, p.cut, r2) %>% 
  group_by(p.cut) %>% summarise(n = n(), r2 = round(sum(r2, na.rm = T), 4))

all.n <- ldl %>% 
  mutate(p.cut = cut(P, breaks = 
                       c(Inf, 0.05, 5e-5, 5e-6, 5e-7, 5e-8, -Inf), lablels = F)) %>% 
  group_by(p.cut) %>% summarise(n = n(), r2 = round(sum(r2, na.rm = T), 4))

tab.ldl <- full_join(all.n, clump.n, by = 'p.cut', suffix = c('.all', '.clump'))
```

```{r alc_cons gwas}

p.ldl <- ggman(filter(ldl, P < 0.05), snp = 'SNP', chrom = 'CHR', bp = 'POS', pvalue = 'P', ymin = 0, ymax = 60, 
               title = "ldl Cholesterol - Willer 2013") + 
  theme_classic() + 
  geom_hline(yintercept = -log10(5e-5), colour = 'blue', size = 0.25) + 
  geom_hline(yintercept = -log10(5e-6), colour = 'red', size = 0.25, linetype = 2)

out.ldl <- grid.arrange(p.ldl,  tableGrob(tab.ldl, theme = ttheme_minimal(base_size = 8)), 
                        ncol = 2, widths = 2:1, layout_matrix = rbind(c(1,2), c(1, NA)))

ggsave('~/Dropbox/Research/PostDoc-MSSM/2_MR/4_Output/ldl.png', device = 'png', units = 'in', width = 10, height = 5, plot = out.ldl)

```

## Total Cholesterol 
[Willer et al Nat Genet 2013](https://www.nature.com/articles/ng.2797): To identify new loci and refine known loci influencing these lipids, we examined 188,577 individuals using genome-wide and custom genotyping arrays. We identify and annotate 157 loci associated with lipid levels at P < 5 × 10−8, including 62 loci not previously associated with lipid levels in humans. 
```{r data_wrangling_exposure, echo=F, cache=TRUE}
TC.raw <- read_table2(params$tc.summary)
TC <- TC.raw %>% 
  rename(SNP = rsid, Effect_allele = A1, Non_Effect_allele = A2, P = `P-value`, EAF = Freq.A1.1000G.EUR, Beta = beta, SE = se) %>% 
  separate(SNP_hg19, c('CHR', 'POS'), sep = ':') %>%
  mutate(CHR = as.numeric(str_replace(CHR, 'chr', ''))) %>% 
  arrange(CHR, POS) %>%
  mutate(Effect_allele = toupper(Effect_allele), Non_Effect_allele = toupper(Non_Effect_allele)) %>%
  filter(nchar(Effect_allele) == 1 & Effect_allele %in% c('A', 'C', 'T', 'G')) %>% 
  filter(nchar(Non_Effect_allele) == 1 & Non_Effect_allele %in% c('A', 'C', 'T', 'G')) %>% 
  mutate(r2 = snp.r2(EAF, Beta)) %>%
  select(SNP, CHR, POS, Effect_allele, Non_Effect_allele, EAF, Beta, SE, P, r2, N)

write_tsv(TC, gzfile('~/Dropbox/Research/PostDoc-MSSM/2_MR/2_DerivedData/tc_GWAS.Processed.gz'))

```

```{r}
## Number of Significant SNPs
TC_ld <- format_data(filter(TC, P < 5e-5), type = 'exposure', 
                      phenotype_col = 'TC',
                      snp_col = 'SNP', 
                      beta_col = "Beta",
                      se_col = "SE",
                      eaf_col = "EAF",
                      effect_allele_col = "Effect_allele",
                      other_allele_col = "Non_Effect_allele",
                      pval_col = "P")
TC_ld <- clump_data(TC_ld, clump_r2 = 0.1, clump_kb = 250)

clump.n <- TC_ld %>% 
  mutate(p.cut = cut(pval.exposure, breaks = 
                       c(Inf, 0.05, 5e-5, 5e-6, 5e-7, 5e-8, -Inf), lablels = F)) %>%
  left_join(select(TC, SNP, r2)) %>% 
  select(SNP, effect_allele.exposure, other_allele.exposure, eaf.exposure, beta.exposure, se.exposure, p.cut, r2) %>% 
  group_by(p.cut) %>% summarise(n = n(), r2 = round(sum(r2, na.rm = T), 4))

all.n <- TC %>% 
  mutate(p.cut = cut(P, breaks = 
                       c(Inf, 0.05, 5e-5, 5e-6, 5e-7, 5e-8, -Inf), lablels = F)) %>% 
  group_by(p.cut) %>% summarise(n = n(), r2 = round(sum(r2, na.rm = T), 4))

tab.TC <- full_join(all.n, clump.n, by = 'p.cut', suffix = c('.all', '.clump'))
```

```{r alc_cons gwas}

p.TC <- ggman(filter(TC, P < 0.05), snp = 'SNP', chrom = 'CHR', bp = 'POS', pvalue = 'P', ymin = 0, ymax = 60, 
               title = "TC Cholesterol - Willer 2013") + 
  theme_classic() + 
  geom_hline(yintercept = -log10(5e-5), colour = 'blue', size = 0.25) + 
  geom_hline(yintercept = -log10(5e-6), colour = 'red', size = 0.25, linetype = 2)

out.TC <- grid.arrange(p.TC,  tableGrob(tab.TC, theme = ttheme_minimal(base_size = 8)), 
                        ncol = 2, widths = 2:1, layout_matrix = rbind(c(1,2), c(1, NA)))

ggsave('~/Dropbox/Research/PostDoc-MSSM/2_MR/4_Output/TC.png', device = 'png', units = 'in', width = 10, height = 5, plot = out.TC)

```


## Triglycerides  
[Willer et al Nat Genet 2013](https://www.nature.com/articles/ng.2797): To identify new loci and refine known loci influencing these lipids, we examined 188,577 individuals using genome-wide and custom genotyping arrays. We identify and annotate 157 loci associated with lipid levels at P < 5 × 10−8, including 62 loci not previously associated with lipid levels in humans. 
```{r data_wrangling_exposure, echo=F, cache=TRUE}
trig.raw <- read_table2(params$tg.summary)
trig <- trig.raw %>% 
  rename(SNP = rsid, Effect_allele = A1, Non_Effect_allele = A2, P = `P-value`, EAF = Freq.A1.1000G.EUR, Beta = beta, SE = se) %>% 
  separate(SNP_hg19, c('CHR', 'POS'), sep = ':') %>%
  mutate(CHR = as.numeric(str_replace(CHR, 'chr', ''))) %>% 
  arrange(CHR, POS) %>%
  mutate(Effect_allele = toupper(Effect_allele), Non_Effect_allele = toupper(Non_Effect_allele)) %>%
  filter(nchar(Effect_allele) == 1 & Effect_allele %in% c('A', 'C', 'T', 'G')) %>% 
  filter(nchar(Non_Effect_allele) == 1 & Non_Effect_allele %in% c('A', 'C', 'T', 'G')) %>% 
  mutate(r2 = snp.r2(EAF, Beta)) %>%
  select(SNP, CHR, POS, Effect_allele, Non_Effect_allele, EAF, Beta, SE, P, r2, N)

write_tsv(trig, gzfile('~/Dropbox/Research/PostDoc-MSSM/2_MR/2_DerivedData/trig_GWAS.Processed.gz'))

```

```{r}
## Number of Significant SNPs
trig_ld <- format_data(filter(trig, P < 5e-5), type = 'exposure', 
                      phenotype_col = 'trig',
                      snp_col = 'SNP', 
                      beta_col = "Beta",
                      se_col = "SE",
                      eaf_col = "EAF",
                      effect_allele_col = "Effect_allele",
                      other_allele_col = "Non_Effect_allele",
                      pval_col = "P")
trig_ld <- clump_data(trig_ld, clump_r2 = 0.1, clump_kb = 250)

clump.n <- trig_ld %>% 
  mutate(p.cut = cut(pval.exposure, breaks = 
                       c(Inf, 0.05, 5e-5, 5e-6, 5e-7, 5e-8, -Inf), lablels = F)) %>%
  left_join(select(trig, SNP, r2)) %>% 
  select(SNP, effect_allele.exposure, other_allele.exposure, eaf.exposure, beta.exposure, se.exposure, p.cut, r2) %>% 
  group_by(p.cut) %>% summarise(n = n(), r2 = round(sum(r2, na.rm = T), 4))

all.n <- trig %>% 
  mutate(p.cut = cut(P, breaks = 
                       c(Inf, 0.05, 5e-5, 5e-6, 5e-7, 5e-8, -Inf), lablels = F)) %>% 
  group_by(p.cut) %>% summarise(n = n(), r2 = round(sum(r2, na.rm = T), 4))

tab.trig <- full_join(all.n, clump.n, by = 'p.cut', suffix = c('.all', '.clump'))
```

```{r alc_cons gwas}

p.trig <- ggman(filter(trig, P < 0.05), snp = 'SNP', chrom = 'CHR', bp = 'POS', pvalue = 'P', ymin = 0, ymax = 80, 
               title = "Triglycerides - Willer 2013") + 
  theme_classic() + 
  geom_hline(yintercept = -log10(5e-5), colour = 'blue', size = 0.25) + 
  geom_hline(yintercept = -log10(5e-6), colour = 'red', size = 0.25, linetype = 2)

out.trig <- grid.arrange(p.trig,  tableGrob(tab.trig, theme = ttheme_minimal(base_size = 8)), 
                        ncol = 2, widths = 2:1, layout_matrix = rbind(c(1,2), c(1, NA)))

ggsave('~/Dropbox/Research/PostDoc-MSSM/2_MR/4_Output/trig.png', device = 'png', units = 'in', width = 10, height = 5, plot = out.trig)

```

## Social Isolation
[Day et al Nat Commu 2018](https://www.nature.com/articles/s41467-018-04930-1): In up to 452,302 UK Biobank study participants, we perform genome-wide association study analyses for loneliness and identify 15 genomic loci (P < 5 × 10−8) for loneliness. 

```{r data_wrangling_exposure, echo=F, cache=TRUE}
sociso.raw <- read_table2(params$sociso.summary)
sociso <- sociso.raw %>% 
  rename(SNP = snpid, CHR = chr, POS = bpos, Effect_allele = a1, Non_Effect_allele = a2, Beta = mtag_beta, SE = mtag_se, EAF = freq, P = mtag_pval, N = n) %>% 
  filter(nchar(Effect_allele) == 1 & Effect_allele %in% c('A', 'C', 'T', 'G')) %>% 
  filter(nchar(Non_Effect_allele) == 1 & Non_Effect_allele %in% c('A', 'C', 'T', 'G')) %>% 
  mutate(r2 = snp.r2(EAF, Beta)) %>%
  select(SNP, CHR, POS, Effect_allele, Non_Effect_allele, EAF, Beta, SE, P, r2, N)

write_tsv(sociso, gzfile('~/Dropbox/Research/PostDoc-MSSM/2_MR/2_DerivedData/sociso_GWAS.Processed.gz'))

```

```{r}
## Number of Significant SNPs
sociso_ld <- format_data(filter(sociso, P < 5e-5), type = 'exposure', 
                       phenotype_col = 'sociso',
                       snp_col = 'SNP', 
                       beta_col = "Beta",
                       se_col = "SE",
                       eaf_col = "EAF",
                       effect_allele_col = "Effect_allele",
                       other_allele_col = "Non_Effect_allele",
                       pval_col = "P")
sociso_ld <- clump_data(sociso_ld, clump_r2 = 0.1, clump_kb = 250)

clump.n <- sociso_ld %>% 
  mutate(p.cut = cut(pval.exposure, breaks = 
                       c(Inf, 0.05, 5e-5, 5e-6, 5e-7, 5e-8, -Inf), lablels = F)) %>%
  left_join(select(sociso, SNP, r2)) %>% 
  select(SNP, effect_allele.exposure, other_allele.exposure, eaf.exposure, beta.exposure, se.exposure, p.cut, r2) %>% 
  group_by(p.cut) %>% summarise(n = n(), r2 = round(sum(r2, na.rm = T), 4))

all.n <- sociso %>% 
  mutate(p.cut = cut(P, breaks = 
                       c(Inf, 0.05, 5e-5, 5e-6, 5e-7, 5e-8, -Inf), lablels = F)) %>% 
  group_by(p.cut) %>% summarise(n = n(), r2 = round(sum(r2, na.rm = T), 4))

tab.sociso <- full_join(all.n, clump.n, by = 'p.cut', suffix = c('.all', '.clump'))
```

```{r alc_cons gwas}

p.sociso <- ggman(filter(sociso, P < 0.05), snp = 'SNP', chrom = 'CHR', bp = 'POS', pvalue = 'P', ymin = 0, ymax = 20, 
                title = "Social Isololation - Day 2018") + 
  theme_classic() + 
  geom_hline(yintercept = -log10(5e-5), colour = 'blue', size = 0.25) + 
  geom_hline(yintercept = -log10(5e-6), colour = 'red', size = 0.25, linetype = 2)

out.sociso <- grid.arrange(p.sociso,  tableGrob(tab.sociso, theme = ttheme_minimal(base_size = 8)), 
                         ncol = 2, widths = 2:1, layout_matrix = rbind(c(1,2), c(1, NA)))

ggsave('~/Dropbox/Research/PostDoc-MSSM/2_MR/4_Output/sociso.png', device = 'png', units = 'in', width = 10, height = 5, plot = out.sociso)

```

## Insomnia 
[Lane et al Nat Genet 2017](https://www.nature.com/articles/ng.3749#supplementary-information): We report single- and multiple-trait genome-wide association analyses of self-reported sleep duration, insomnia symptoms and excessive daytime sleepiness in the UK Biobank (n = 112,586).
```{r data_wrangling_exposure, echo=F, cache=TRUE}
insom.raw <- read_table2(params$insom.summary)
insom <- insom.raw %>% 
  rename(POS = BP, Effect_allele = A1, Non_Effect_allele = A2, EAF = MAF) %>% 
  filter(nchar(Effect_allele) == 1 & Effect_allele %in% c('A', 'C', 'T', 'G')) %>% 
  filter(nchar(Non_Effect_allele) == 1 & Non_Effect_allele %in% c('A', 'C', 'T', 'G')) %>% 
  filter(CHR %in% 1:22) %>%
  arrange(CHR, POS) %>% 
  mutate(Beta = log(OR)) %>% 
  mutate(r2 = snp.r2(EAF, Beta)) %>%
  select(SNP, CHR, POS, Effect_allele, Non_Effect_allele, EAF, Beta, SE, P, r2, N)

write_tsv(insom, gzfile('~/Dropbox/Research/PostDoc-MSSM/2_MR/2_DerivedData/insom_GWAS.Processed.gz'))

```

```{r}
## Number of Significant SNPs
insom_ld <- format_data(filter(insom, P < 5e-5), type = 'exposure', 
                       phenotype_col = 'insom',
                       snp_col = 'SNP', 
                       beta_col = "Beta",
                       se_col = "SE",
                       eaf_col = "EAF",
                       effect_allele_col = "Effect_allele",
                       other_allele_col = "Non_Effect_allele",
                       pval_col = "P")
insom_ld <- clump_data(insom_ld, clump_r2 = 0.1, clump_kb = 250)

clump.n <- insom_ld %>% 
  mutate(p.cut = cut(pval.exposure, breaks = 
                       c(Inf, 0.05, 5e-5, 5e-6, 5e-7, 5e-8, -Inf), lablels = F)) %>%
  left_join(select(insom, SNP, r2)) %>% 
  select(SNP, effect_allele.exposure, other_allele.exposure, eaf.exposure, beta.exposure, se.exposure, p.cut, r2) %>% 
  group_by(p.cut) %>% summarise(n = n(), r2 = round(sum(r2, na.rm = T), 4))

all.n <- insom %>% 
  mutate(p.cut = cut(P, breaks = 
                       c(Inf, 0.05, 5e-5, 5e-6, 5e-7, 5e-8, -Inf), lablels = F)) %>% 
  group_by(p.cut) %>% summarise(n = n(), r2 = round(sum(r2, na.rm = T), 4))

tab.insom <- full_join(all.n, clump.n, by = 'p.cut', suffix = c('.all', '.clump'))
```

```{r alc_cons gwas}

p.insom <- ggman(filter(insom, P < 0.05), snp = 'SNP', chrom = 'CHR', bp = 'POS', pvalue = 'P', ymin = 0, ymax = 20, 
                title = "Insomnia - Lane 2017") + 
  theme_classic() + 
  geom_hline(yintercept = -log10(5e-5), colour = 'blue', size = 0.25) + 
  geom_hline(yintercept = -log10(5e-6), colour = 'red', size = 0.25, linetype = 2)

out.insom <- grid.arrange(p.insom,  tableGrob(tab.insom, theme = ttheme_minimal(base_size = 8)), 
                         ncol = 2, widths = 2:1, layout_matrix = rbind(c(1,2), c(1, NA)))

ggsave('~/Dropbox/Research/PostDoc-MSSM/2_MR/4_Output/insom.png', device = 'png', units = 'in', width = 10, height = 5, plot = out.insom)

```

## Sleep Duration 
[Lane et al Nat Genet 2017](https://www.nature.com/articles/ng.3749#supplementary-information): We report single- and multiple-trait genome-wide association analyses of self-reported sleep duration, insomnia symptoms and excessive daytime sleepiness in the UK Biobank (n = 112,586).
```{r data_wrangling_exposure, echo=F, cache=TRUE}
sleep.raw <- read_table2(params$sleep.summary)
sleep <- sleep.raw %>% 
  rename(POS = BP, Effect_allele = A1, Non_Effect_allele = A2, EAF = MAF, Beta = BETA) %>% 
  filter(nchar(Effect_allele) == 1 & Effect_allele %in% c('A', 'C', 'T', 'G')) %>% 
  filter(nchar(Non_Effect_allele) == 1 & Non_Effect_allele %in% c('A', 'C', 'T', 'G')) %>% 
  filter(CHR %in% 1:22) %>%
  arrange(CHR, POS) %>% 
  mutate(r2 = snp.r2(EAF, Beta)) %>%
  select(SNP, CHR, POS, Effect_allele, Non_Effect_allele, EAF, Beta, SE, P, r2, N)

write_tsv(sleep, gzfile('~/Dropbox/Research/PostDoc-MSSM/2_MR/2_DerivedData/sleep_GWAS.Processed.gz'))

```

```{r}
## Number of Significant SNPs
sleep_ld <- format_data(filter(sleep, P < 5e-5), type = 'exposure', 
                       phenotype_col = 'sleep',
                       snp_col = 'SNP', 
                       beta_col = "Beta",
                       se_col = "SE",
                       eaf_col = "EAF",
                       effect_allele_col = "Effect_allele",
                       other_allele_col = "Non_Effect_allele",
                       pval_col = "P")
sleep_ld <- clump_data(sleep_ld, clump_r2 = 0.1, clump_kb = 250)

clump.n <- sleep_ld %>% 
  mutate(p.cut = cut(pval.exposure, breaks = 
                       c(Inf, 0.05, 5e-5, 5e-6, 5e-7, 5e-8, -Inf), lablels = F)) %>%
  left_join(select(sleep, SNP, r2)) %>% 
  select(SNP, effect_allele.exposure, other_allele.exposure, eaf.exposure, beta.exposure, se.exposure, p.cut, r2) %>% 
  group_by(p.cut) %>% summarise(n = n(), r2 = round(sum(r2, na.rm = T), 4))

all.n <- sleep %>% 
  mutate(p.cut = cut(P, breaks = 
                       c(Inf, 0.05, 5e-5, 5e-6, 5e-7, 5e-8, -Inf), lablels = F)) %>% 
  group_by(p.cut) %>% summarise(n = n(), r2 = round(sum(r2, na.rm = T), 4))

tab.sleep <- full_join(all.n, clump.n, by = 'p.cut', suffix = c('.all', '.clump'))
```

```{r alc_cons gwas}

p.sleep <- ggman(filter(sleep, P < 0.05), snp = 'SNP', chrom = 'CHR', bp = 'POS', pvalue = 'P', ymin = 0, ymax = 20, 
                title = "Sleep Duration - Lane 2017") + 
  theme_classic() + 
  geom_hline(yintercept = -log10(5e-5), colour = 'blue', size = 0.25) + 
  geom_hline(yintercept = -log10(5e-6), colour = 'red', size = 0.25, linetype = 2)

out.sleep <- grid.arrange(p.sleep,  tableGrob(tab.sleep, theme = ttheme_minimal(base_size = 8)), 
                         ncol = 2, widths = 2:1, layout_matrix = rbind(c(1,2), c(1, NA)))

ggsave('~/Dropbox/Research/PostDoc-MSSM/2_MR/4_Output/sleep.png', device = 'png', units = 'in', width = 10, height = 5, plot = out.sleep)

```


## Personality 
[Lo et al Nat Genet 2017](https://dx.doi.org/10.1038%2Fng.3736) We identified six genetic loci, including five novel loci2,3, significantly associated with personality traits in a meta-analysis of genome-wide association studies (N=123,132–260,861). Summary statistics of GWAS from 23andMe (available in Supplementary Data Sets 1–5 for the top 10K SNPs) were combined with the two GPC samples separately.
```{r data_wrangling_exposure, echo=F, cache=TRUE}
agree.raw <- read_table2(params$agree.summary, guess_max = 10000)
agree <- agree.raw %>% 
  rename(SNP = MarkerName, Effect_allele = A1, Non_Effect_allele = A2, P = Pval) %>% 
  filter(nchar(Effect_allele) == 1 & Effect_allele %in% c('A', 'C', 'T', 'G')) %>% 
  filter(nchar(Non_Effect_allele) == 1 & Non_Effect_allele %in% c('A', 'C', 'T', 'G')) %>% 
  mutate(r2 = snp.r2(EAF, Beta)) %>%
  mutate(N = 1131881) %>% 
  select(SNP, CHR, POS, Effect_allele, Non_Effect_allele, EAF, Beta, SE, P, r2, N)

write_tsv(agree, gzfile('~/Dropbox/Research/PostDoc-MSSM/2_MR/2_DerivedData/agree_GWAS.Processed.gz'))

```

```{r}
## Number of Significant SNPs
agree_ld <- format_data(filter(agree, P < 5e-5), type = 'exposure', 
                       phenotype_col = 'agree',
                       snp_col = 'SNP', 
                       beta_col = "Beta",
                       se_col = "SE",
                       eaf_col = "EAF",
                       effect_allele_col = "Effect_allele",
                       other_allele_col = "Non_Effect_allele",
                       pval_col = "P")
agree_ld <- clump_data(agree_ld, clump_r2 = 0.1, clump_kb = 250)

clump.n <- agree_ld %>% 
  mutate(p.cut = cut(pval.exposure, breaks = 
                       c(Inf, 0.05, 5e-5, 5e-6, 5e-7, 5e-8, -Inf), lablels = F)) %>%
  left_join(select(agree, SNP, r2)) %>% 
  select(SNP, effect_allele.exposure, other_allele.exposure, eaf.exposure, beta.exposure, se.exposure, p.cut, r2) %>% 
  group_by(p.cut) %>% summarise(n = n(), r2 = round(sum(r2, na.rm = T), 4))

all.n <- agree %>% 
  mutate(p.cut = cut(P, breaks = 
                       c(Inf, 0.05, 5e-5, 5e-6, 5e-7, 5e-8, -Inf), lablels = F)) %>% 
  group_by(p.cut) %>% summarise(n = n(), r2 = round(sum(r2, na.rm = T), 4))

tab.agree <- full_join(all.n, clump.n, by = 'p.cut', suffix = c('.all', '.clump'))
```

```{r alc_cons gwas}

p.agree <- ggman(filter(agree, P < 0.05), snp = 'SNP', chrom = 'CHR', bp = 'POS', pvalue = 'P', ymin = 0, ymax = 60, 
                title = "Agreebleness - Lo 2017") + 
  theme_classic() + 
  geom_hline(yintercept = -log10(5e-5), colour = 'blue', size = 0.25) + 
  geom_hline(yintercept = -log10(5e-6), colour = 'red', size = 0.25, linetype = 2)

out.agree <- grid.arrange(p.TRAIT,  tableGrob(tab.TRAIT, theme = ttheme_minimal(base_size = 8)), 
                         ncol = 2, widths = 2:1, layout_matrix = rbind(c(1,2), c(1, NA)))

ggsave('~/Dropbox/Research/PostDoc-MSSM/2_MR/4_Output/TRAIT.png', device = 'png', units = 'in', width = 10, height = 5, plot = out.TRAIT)

```

## IBD 

```{r data_wrangling_exposure, echo=F, cache=TRUE}
ibd.raw <- read_table2(params$ibd.summary)
ibd <- ibd.raw %>% 
  separate(MarkerName, c('CHR', 'POS'), sep = ':', remove = F) %>% 
  separate('POS', c('POS', 'REF', 'ALT'), sep = '_') %>% 
  rename(SNP = MarkerName, Effect_allele = Allele1, Non_Effect_allele = Allele2, Beta = Effect, SE = StdErr, P = P.value) %>% 
  filter(nchar(Effect_allele) == 1 & Effect_allele %in% c('A', 'C', 'T', 'G')) %>% 
  filter(nchar(Non_Effect_allele) == 1 & Non_Effect_allele %in% c('A', 'C', 'T', 'G')) %>% 
  mutate(Effect_allele = toupper(Effect_allele), Non_Effect_allele = toupper(Non_Effect_allele)) %>%
  left_join(select(hrc, ID, POS, REF, ALT, AF), by = c('SNP' = 'ID')) %>%
  mutate(EAF = ifelse(Effect_allele == ALT, AF, 1-AF)) %>% 
  mutate(r2 = snp.r2(EAF, Beta)) %>%
  mutate(N = 59957) %>% 
  select(SNP, CHR, POS, Effect_allele, Non_Effect_allele, EAF, Beta, SE, P, r2, N)

write_tsv(ibd, gzfile('~/Dropbox/Research/PostDoc-MSSM/2_MR/2_DerivedData/ibd_GWAS.Processed.gz'))

```

```{r}
## Number of Significant SNPs
ibd_ld <- format_data(filter(ibd, P < 5e-5), type = 'exposure', 
                       phenotype_col = 'ibd',
                       snp_col = 'SNP', 
                       beta_col = "Beta",
                       se_col = "SE",
                       eaf_col = "EAF",
                       effect_allele_col = "Effect_allele",
                       other_allele_col = "Non_Effect_allele",
                       pval_col = "P")
ibd_ld <- clump_data(ibd_ld, clump_r2 = 0.1, clump_kb = 250)

clump.n <- ibd_ld %>% 
  mutate(p.cut = cut(pval.exposure, breaks = 
                       c(Inf, 0.05, 5e-5, 5e-6, 5e-7, 5e-8, -Inf), lablels = F)) %>%
  left_join(select(ibd, SNP, r2)) %>% 
  select(SNP, effect_allele.exposure, other_allele.exposure, eaf.exposure, beta.exposure, se.exposure, p.cut, r2) %>% 
  group_by(p.cut) %>% summarise(n = n(), r2 = round(sum(r2, na.rm = T), 4))

all.n <- ibd %>% 
  mutate(p.cut = cut(P, breaks = 
                       c(Inf, 0.05, 5e-5, 5e-6, 5e-7, 5e-8, -Inf), lablels = F)) %>% 
  group_by(p.cut) %>% summarise(n = n(), r2 = round(sum(r2, na.rm = T), 4))

tab.ibd <- full_join(all.n, clump.n, by = 'p.cut', suffix = c('.all', '.clump'))
```

```{r alc_cons gwas}

p.ibd <- ggman(filter(ibd, P < 0.05), snp = 'SNP', chrom = 'CHR', bp = 'POS', pvalue = 'P', ymin = 0, ymax = 60, 
                title = "ibd - AUTHOR") + 
  theme_classic() + 
  geom_hline(yintercept = -log10(5e-5), colour = 'blue', size = 0.25) + 
  geom_hline(yintercept = -log10(5e-6), colour = 'red', size = 0.25, linetype = 2)

out.ibd <- grid.arrange(p.ibd,  tableGrob(tab.ibd, theme = ttheme_minimal(base_size = 8)), 
                         ncol = 2, widths = 2:1, layout_matrix = rbind(c(1,2), c(1, NA)))

ggsave('~/Dropbox/Research/PostDoc-MSSM/2_MR/4_Output/ibd.png', device = 'png', units = 'in', width = 10, height = 5, plot = out.ibd)

```







